# Introduction to the tidyverse

## What on earth is the tidyerse?

Broadly and briefly, the tidyverse is a collection of packages
that are tailored to operating on data that are stored in what
is called a "tidy" format.  Such a format is one that is typical
of the manner in which data are stored in relational data bases.
(We will be talking more about "tidy" format throughout this session).

There is an R package called 'tidyverse' that is simply a wrapper that
automatically loads the 8 _core_ packages of the tidyverse. In this session
we will talk specifically about functions in 5 of those packages, encountering
them in this order:

- 'readr'
- 'tibble'
- 'tidyr'
- 'dplyr'
- 'purrr'

The other core packages in the tidyverse are:

- 'ggplot2' --- the well known graphics plotting package
- 'stringr' --- a package for manipulating string data
- 'forcats' --- a package to simplify the use of factors

These packages provide a unified way of handling data in a tidy format.
And, it turns out, almost any data type can be stored and expressed
in a tidy format.  This means that 99% of your data analysis tasks
can be tackled in a tidy fashion.  

Some "big-data" data sets (like
genome sequencing alignments, or doppler radar data, etc.) come in
specialized data formats that allow for fast access and compression,
and would not by amenable to storage in a tidy format.  However, these
are specialized cases, and for almost every data analysis application
you will encounter, it behooves you to get comfortable in the
tidyverse.

At any rate, to have access to all the functions from the 8 core
packages of the 'tidyverse' we just load the package like this:
```{r}
library(tidyverse)
```

## An example data set

We will start our exploration of the tidyverse with a data set
that is decidedly _not_ tidy.  It is a small data set of _genotypes_
of 34 fish from four different rivers with code names `BALGA`, `BCHNC`, 
`BEMME`, and `BMTMA`.  The fish have been genotyped
at 5 genetic markers, named `mk1`, `mk2`, `mk3`, `mk4`, and `mk5`.  
Because these fish are _diploids_ they have two copies of each
gene, and therefore, their genotype at a single markers
consists of the _allele_ or _allelic type_ of the two gene copies
at that marker.  So, in this data set, for example, the `mk1_1` column
holds information about the first gene copy at marker `mk1` and,
`mk1_2` holds information about the second gene copy at marker `mk1`.
Different alleles are specified as different positive integer
values, (like `228`) 

The data set is in the merida-workshop-2022 Rstudio project, stored as a CSV
file at `data/tidy-intro/genotypes.csv`.  It looks like this:
```csv
`r paste(readLines("data/tidy-intro/genotypes.csv"), collapse = "\n")`
```

## Reading data in the tidyverse

The package 'readr' has a large number of functions for reading in
data of different types.  They all start with `read_`.  These functions
typically read things into tidyverse-specific data frames that are
called "tibbles," about which we will hear more later.

We can read the `genotypes.csv` file using the readr function,
`read_csv()`.  

Note: this looks like the base R function `read.csv()`, but it is not the
same!!  Most notably, 

- `read_csv()` will never automatically convert strings to factors (Rejoice!)
- `read_csv()` is faster than `read.csv()`
- `read_csv()` is more likely to tell you if there are unexpected
irregularities found while reading your data set.  It also has 
facilities to tell you exactly where those data-reading problems
occurred.
- `read_csv()` will return data in a _tibble_ rather than simply
a _data frame_.  (More about this soon!).

These characteristics are shared amongst all the `read_*()` functions
(such as `read_tsv()`, `read_table()`, `read_table2()` and 
`read_delim()`) in the 'readr' package.

A delightful overview of the readr package can be found on the
"readr Cheatsheet" from RStudio available 
[here](https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-import.pdf) or
by clicking the thumbnail below.

[![](https://github.com/rstudio/cheatsheets/raw/main/pngs/thumbnails/data-import-cheatsheet-thumbs.png)](https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-import.pdf)

Reading the data set with `read_csv()` is simple:
```{r, message=FALSE}
genos <- read_csv("data/tidy-intro/genotypes.csv")
```
And then we can display it like this:
```{r}
genos
```
Each row coresponds to one individual, and each column corresponds to
one of the two gene copies at a genetic marker.

## A word on tibbles

It is worth noting how it is that this tibble, this special type
of a _data frame_ is displayed when you print it:

- The first line, `# A tibble: 34 Ã— 12`,  tells how many rows (34) and
columns (12) are in the tibble.
- By default, only the first ten rows are printed.
- Beneath the name of each column, the _type_ of data in the column
is listed.  In this case, `<dbl>` means numeric data and `<chr>` means
the column is a vector of character (string) data (i.e. text). Other possibilities
are `<int>` for integer, `<fct>` for factor, and `<list>` for a column that
is a list (Richard will talk more about that!)
- If there are too many columns for the width of the screen, only
the contents of the first few columns are printed. The remaining columns
are just shown beneath by their names and types.
- If there a a very large number of columns, the names and types of some of
the remaining ones are not shown.

In this way, you can quickly get a sense for how a data set is
structured by simply printing it to the screen without running the
risk of bombing your whole R console by having it print all the lines
of a very large data frame!

While the `readr` package always returns a tibble, the `as_tibble()`
function from the 'tibble' package provides an easy way to turn
an existing data frame into a tibble.  For example, to witness
how much easier it is to see what is contained in the built-in
R data set, `iris`, by converting it into a tibble, the reader
is invited to evaluate the following two lines in their
R console and compare the output:
```{r, eval=FALSE}
iris # the whole data set gets spit out and you end up at bottom of it

as_tibble(iris) # a nice tibble summary gets printed
```


Finally, the 'tibble' package provides the function `tibble()` that
allows the user to create a tibble from vectors.  It is much like
the built-in R function, `data.frame()` except that it deals appropriately
with list vectors.  

For more about tibbles, the best place if you are new is the
[tibble chapter](https://r4ds.had.co.nz/tibbles.html), in the _R For Data Science_
book.


## Back to "tidying" up our genotype data

Our small data set `genos` might seem pretty neat and tidy---it is easy
to look at as a human, and it would not take a whole lot of space to
print it out on a page.  However, in the current format, doing
a number of summaries of the data would require somewhat specialized
code.

For example, if you wanted to count up the number of occurrences of each
allelic type (i.e. each number, like `311`) at each of the markers, within
each of the rivers, it would not be straightforward! Just imagine
counting alleles at markers `mk_1`: the alleles are in two different
columns, and you would have to account for fish being from different rivers.

While one could certainly write base R code to traverse the data frame and
pick out the appropriate columns and rows to make these summaries, the
_tidyverse_ approach is to recognize that converting the data to
a different format (often called a "long" format), will make it easier
to do most any operation on the data.

The main problem of the current data set, `genos`, is that
the observed values at all the markers are the same---they are all
just different _alleles_; however, they occupy lots of different columns.
A general principle in tidying data is to strive for having only a single
column to represent a certain type of observation.

In the case of `genos`,
the column headers `mk1_1, mk1_2, ... , mk_5_2` tell us which particular
markers and gene copies have a particular allelic type.  But another way
to represent that is to have a column that gives the marker and the gene copy.

Perhaps it is best to simply show what a tidy, long version of these data
look like.  The core tidyverse package 'tidyr' provides many functions for
tidying data sets.  We will use the `pivot_longer()` function to turn
the column names into data values in new columns:
```{r}
genos_long <- genos %>%
  pivot_longer(
    cols = c(mk1_1:mk5_2), 
    names_to = c("marker", "gene_copy"),
    names_sep = "_", 
    values_to = "allele"
  )
genos_long
```

Wow, that is a wildly different format!  We now only have 5 columns:
  
- `river`: tells us what river the fish is from
- `indiv`: the ID of the individual fish
- `marker`: the name of the marker (i.e. `mk_1`, or `mk_2`, etc.)
- `gene_copy`: whether we are looking at the first (`1`) or the second (`2`)
  gene copy of the fish.
- `allele`: the allelic type of the gene_copy at the marker in the individual fish from
  the particular river.
  
Though our data set has only 5 columns, it now has 340 rows.  As a consequence,
it is not really possible to "look at" a large portion of the data set as a human
(as was possible before).

Furthermore, if you don't know how to handle data like these
it can seem daunting.  But, once you learn the tools of the tidyverse (and particularly
of the package 'dplyr') this format lets you learn the same, simple approach that
will work for many, many different data sets.

The 'tidyr' package has a number of functions for tidying data.  Check out
the [RStudio tidyr cheatsheet](https://raw.githubusercontent.com/rstudio/cheatsheets/main/tidyr.pdf) for a great overview of 'tidyr' features.

[![](https://raw.githubusercontent.com/rstudio/cheatsheets/master/pngs/thumbnails/tidyr-thumbs.png)](https://raw.githubusercontent.com/rstudio/cheatsheets/main/tidyr.pdf)

## What the heck is that `%>%` thingie?!

Before we proceed, we want to make sure that everyone understands
what the `%>%` in the above code block is doing.

The packages of the tidyverse all import the R symbol `%>%` from the
'magrittr' package.  This is referred to as the tidyverse "pipe" because
it allows the user to "pipe" the output of one function into the input
for another function (in the same way that experience Unix computer users
will "pipe" output from one command into the input for another command).

In short, the `%>%` symbol takes whatever is on the
_left_ of it and it supplies that as the first argument of the function
that is to the _right_ of it.  The thing on the left can be an R object or
the output of a functio.

For a simple, contrived example, suppose you wanted to add the integers from 
1 to 10, take the square root of the result, and then calculate the sin of that
result, and then take the log of it.  Without piping capability, you could write that
operation in a horribly nested set of parentheses:
```{r}
log(sin(sqrt(sum(1:10))))
```
which is hard to read because the direction of operations runs
right to left.  It is much easier to read and understand what is going
on by piping results from one function to the next:
```{r}
1:10 %>%
  sum() %>%
  sqrt() %>%
  sin() %>%
  log()
```

Operating upon tibbles with the pipe let's you chain multiple
different functions/operations upon the data in an easy-to-read
fashion. And, it let's you check intermediate results along the way
as you build up more complex expressions.  It is a crucial
capability in the tidyverse.

We will see the tidyverse pipe used extensively in the following.


## The 'dplyr' package: the heart of the tidyverse

'dplyr' is the tidyverse package that offers the most functionality for operating
on data sets.  The main 'dplyr' functions for performing actions on a tibble
are described as "verbs": and their names are meant to describe the actions they do.

There are a lot of other functions within 'dplyr'. We can't cover them all here,
but, once again, the 
[RStudio dplyr cheatsheet](https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-transformation.pdf) is the perfect place for an overview.

[![](https://raw.githubusercontent.com/rstudio/cheatsheets/master/pngs/thumbnails/data-transformation-cheatsheet-thumbs.png)](https://raw.githubusercontent.com/rstudio/cheatsheets/main/data-transformation.pdf)



Here, we will cover just the main verbs, namely:

- `select()`, `filter()`, `arrange()` change the extent of columns, or rows, or
the sort order of rows, respectively.

- `mutate()` allows for the creation of new columns that are functions of
existing columns (and have the same length as the existing columns).

- `summarise()` allows for summarising existing columns down to a single value.

- `group_by()` allows the 'dplyr' verbs to operate on multiple subsets of data
defined by the values of grouping variables.  I dare say this is the most
important concept in all the tidyverse.

- `left_join()` the simplest of a whole family of _joins_ (`left_join()`, `full_join()`,
`inner_join()`, etc.), that
combine rows of two tibbles together according to shared values in certain columns.  


### Non-standard Evaluation (NSE): You don't wrap column names in quotation marks

Importantly, in the tidyverse, when you operate on a column of a tibble, it is
customary to refer to that column by its _name_, rather than by the number that
describes its position (this makes the code more resilient to things like ordering
of columns in a tibble).  Within the tidyverse, names of columns are given
as if they were variable names---they should almost never be wrapped
in quotation marks.  

### `select()` to select columns of a tibble

The select function is used to choose which columns of a tibble are returned.
They get returned in a tibble of their own.  Examples:

```{r}
# return a tibble with just the river and indiv of
# the original genos tibble
genos %>%
  select(river, indiv)
```
Ranges of columns can be selected by putting a colon between
the name of the column at the start of the range and the name
of the column at the end of the range.  

Furthermore, adding a `-` to a column name in `select()` will remove
it.  And adding a `-` before a column name range (surrounded by parentheses) will
remove all those columns.  For example:
```{r}
# remove the columns mk1_1 to mk3_2 from the original
# genos tibble
genos %>%
  select(-(mk1_1:mk3_2))
```

### `filter()` to retain certain rows

The `filter()` function takes any vectorized logical expression involving
columns of the tibble.  Rows for which that expression evaluate to `TRUE`
are retained, and those that evaluate to `FALSE` are not.

Examples:
```{r}
# only keep rows with river code "BALGA"
genos_long %>%
  filter(river == "BALGA")
```

```{r}
# only keep data from markers mk3 and mk4
genos_long %>%
  filter(marker %in% c("mk3", "mk4"))
```
```{r}
# remove rows for which allele is NA
genos_long %>%
  filter(!is.na(allele))
```

### `arrange()` to reorder the rows of the tibble

For example, perhaps you wish to see which of the alleles have the
highest numbers.  Then:
```{r}
genos_long %>%
  arrange(desc(allele))
```
The 'dplyr' function `desc()` means: arrange rows in descending order of
the variable named within it.  

You can arrange/sort rows according to multiple columns, as well.  For example 
to sort things that show individual genotypes in adjacent rows,
but with rivers in reverse alphabetical order, we could do:
```{r}
genos_long %>%
  arrange(desc(river), indiv, marker, gene_copy)
```

### `mutate()` to create new columns

With the `mutate()` function, it is easy to create a new
column that is a function of other columns.  The syntax is like:
```{r, eval=FALSE}
# create a new column named new_column that is a function
# or existing columns col1, col2
genos_long %>%
  mutate(new_column_name = some_function(col1, col2))
``` 

In this capacity, 'dplyr' is very particular:

* If the function returns the same number of rows as the tibble,
it works.
* If the function returns a single value, that single value is
replicated in each row of the new column.
* Otherwise, `mutate()` throws an error.

Simple examples.  Try these on your own:
```{r, eval=FALSE}
# make a column, "long_marker_names" that give
# a long name in the form "Marker X" for each marker
genos_long %>%
  mutate(long_marker_names = str_replace(marker, "mk", "Marker "))

# make a new column called gc_ab that gives the gene copies
# as "a" and "b", rather than "1" and "2"
genos_long %>%
  mutate(
    gc_ab = case_when(
      gene_copy == 1 ~ "a",
      gene_copy == 2 ~ "b",
      TRUE ~ NA_character_
    )
  )

# make a new column called season that tells us that
# all of these data were collected in the "Autumn"
genos_long %>%
  mutate(season = "Autumn")
```
### Compute a single value from many rows: `summarise()`

While `mutate()` always returns a new column that is the same length
as all the other columns in the tibble, `summarise()` does something
quite different: it allows you to make a column that will have only
a single row in it.  We will see more examples later on, but for now,
consider one of the simplest ways to summarise a tibble: count
how many rows are in it.  That is done with the 'dplyr' function
`n()`:
```{r}
genos_long %>%
  summarise(num_rows = n())
```
See how that `summarise()` function takes an entire tibble, and returns
a single value (the number of rows of the original tibble).

This behavior might not seem all that useful, but as we will see,
when combined with _grouping_ it is extraordinarily useful.

### `group_by()` to operate tidily on subsets of a tibble

One of the most important aspects of working with tibbles is
that of  _grouping_ them by variables.  In fact, without _grouping_
the "tidy" or "long" format would be nearly impossible to deal with; however
with grouping, a huge number of problems can be solved upon tidy data
within the tidyverse.

Much of the time, we want to apply functions to separate parts of the
data set, each part in turn.  An example might help: suppose that
we want to compute the allele frequencies from `genos_long`. In other words,
we want to count how many times _within each river_, _each allele_, _at each locus_ is observed.  

Each time we say something like _at each locus_, or 
_within each river_, we are implicitly talking about _grouping_ our
data. Thus, if we "group our data by river," we are conceptually,
breaking our data set into four distinct tibbles:

- one with all the rows in which `river == "BALGA"`
- one with all the rows in which `river == "BCHNC"`
- one with all the rows in which `river == "BEMME"`
- one with all the rows in which `river == "BMTMA"`

Then (and this is the _really_ important part!) if we
pass such a "grouped" tibble to `mutate()` or `summarise()`,
then those functions will operate on each of those four separate
groups, independently, but it will still return the result
as a single tibble.

To group a tibble, we use the `group_by()` function.  Here we
group `genos_long` by `river`:
```{r}
genos_long %>%
  group_by(river)
```
From the output that gets printed out, it looks like almost nothing
has changed about the tibble, _except_ there is this extra line:
```
# Groups:   river [4]
```
This tells us that after running `genos_long` through `group_by(river)`,
the result is a tibble that is grouped by the `river` column, and there
are 4 groups.

But, now, if we count the number of rows in each of those groups
we start to see how powerful this approach can be:
```{r}
genos_long %>%
  group_by(river) %>%
  summarise(num_rows = n())
```

OK! That is telling us about the sizes of those 4 groups, and we can
conceptually think of each group as being a "sub-tibble" with the
number of rows as listed above.

Note that you can group tibbles by more than one column.  When grouped on
multiple columns, the groups are formed from all the observed combinations
of values in the different columns.  For example:
```{r}
genos_long %>%
  group_by(river, marker, allele)
```
apparently groups the data into 93 different groups.

Now, we are ready to try computing allele frequencies (i.e. counting alleles) at different
markers.  Remember, we want to count how many times _within each river_ and _at each locus_ that
_each allele_ is observed.  (Note we also don't want to count missing data for this so we can 
filter it out in the beginning).  Hence:
```{r}
genos_long %>%
  filter(!is.na(allele)) %>%
  group_by(river, marker, allele) %>%
  summarise(num_alleles = n())
```

Wow! That is fast and easy.  

It is relatively straightforward to think about how `summarise()` works on grouped tibbles, but
it is just as important to know that `mutate()` also operates on the separate groups of a
grouped tibble.  

Note, that the result above has the notation:
```
# Groups:   river, marker [20]
```
So, the result is still a grouped tibble, but in this case it is grouped by
`river, marker`.  We had originally grouped it by `river, marker, allele`, BUT
by default, the `summarise()` function will return a tibble in which the last
grouping variable (in this case `allele`) is no longer in effect.

In our case, this is convenient, because such a grouping is good for
computing _relative frequencies_ of the alleles.  We can just add that on
to the chain with the pipe.  (And while we are at it, sort things
by allele frequency.)
```{r}
genos_long %>%
  filter(!is.na(allele)) %>%
  group_by(river, marker, allele) %>%
  summarise(num_alleles = n()) %>%
  mutate(freq = num_alleles / sum(num_alleles)) %>%
  ungroup() %>% 
  arrange(river, marker, desc(freq))

```

A few things are worth noting:

- `sum(num_allele)` sums the number of alleles in each river at each marker (because
that is how the tibble is grouped).  Hence, `num_allele / sum(num_allele)` gives the
relative frequency of each allele (in the river at the marker).
- `ungroup()` removes any grouping criteria from a tibble.  It is handy if you
don't want to have any groups in your tibble any longer.
- The arrange function lets us sort the rows in a useful fashion.

#### Exercises for you

The reason that the tidy format is wonderful is because you can get a number
of different results by following the same principles, but grouping things
differently.

Try using the tools of 'dplyr' to compute the following.
A skeleton has been provided.  Put the missing code into the `...` in each.

1. A tibble with the relative frequency of missing data within each river
```{r, eval=FALSE}
genos_long %>%
  group_by(...) %>%
  summarise(fract_miss = sum(is.na(allele) / n())
```

2. A tibble with the relative frequency of missing data at each marker
```{r, eval=FALSE}
genos_long %>%
  group_by(...) %>%
  summarise(fract_miss = ...)
```

### Joins

Joins are what you use to add columns from one tibble to
another tibble by matching the values in one or more shared columns.

We will just show the simplest join, the `left_join()`.

Suppose that we have data about each fish, that tells us:

- whether they were caught in the autumn or the spring
- whether they were a juvenile or an adult
- the length in centimeters.

We have an example of such a data set that we can read in:
```{r, message=FALSE}
fish_meta <- read_csv("data/tidy-intro/fish-metadata.csv")
fish_meta
```

OK! This has season, stage, and length for 1,252 fish.  That is a lot more
than the 34 fish we have genotype data for.  So, how can we pick out the relevant
rows from `fish_meta` to join them on to the columns of `genos_long`? We use
`left_join()`.  

Note that in both `genos_long` and in `fish_meta`, the individual ID is
in the column `indiv`.  This means that we can _join_ the two tibbles
_by_ the `indiv` column in each tibble.  Thus:
```{r}
genos_long2 <- genos_long %>%
  left_join(fish_meta, by = "indiv")
genos_long2
```

Now, you could do all sorts of things like explore allele frequency differences
in different seasons or from different life stages.

### Wrap Up

We have just scratched the surface of what is possible with the tidyverse.

One thing we haven't mentioned much is that the columns of a tibble can, themselves,
be list-vectors.  This can be very powerful when doing statistical modeling, as the
outputs of different models can be stored in lists that are themselves columns of
tibbles.  Manipulating such lists is done with the `purrr` package, which is part of
the tidyverse. 

Richard will give us an example of that for the rest of the session.


## Modeling within the Tidyverse
Load required packages

```{r}
library(viridis) # preferred color palette for graphing 
```

Import bird data.

The data provides information on the species richness of birds and butterflies and NDVI at 30 sites.
Our goal is to run a linear model that estimates the magnitude, direction, and uncertainty in the species richness-NDVI relationship for each taxa and then visualize the relationship.

```{r}
data_all <- read_csv("data/data.richard.new.csv")

data_all
```

The old way is to run a model on each taxa separately:

```{r}
# subset the Birds data
model_birds <- lm(Richness ~ NDVI, data = data_all[data_all$Taxa == "Birds", ]) 

# subset the Butterflies
model_butterflies <- lm(Richness ~ NDVI, data = data_all[data_all$Taxa == "Butterflies", ]) 
```

Extracting results from the summary file is not easy.

```{r}
# You have to remember that subscript [[4]] corresponds to the main results

key_results_birds <- as.data.frame(summary(model_birds)[[4]]) 

```

The broom package presents model results in a tidy way.

```{r}
library(broom)
```

broom::tidy() returns parameter estimates.

```{r}
birds_tidy <- model_birds %>% 
  tidy()

birds_tidy
```

broom::glance() includes R2, AIC, and more.

```{r}
birds_glance <- model_birds %>% 
  glance()

birds_glance
```

broom::augment() returns fitted values

```{r}
birds_augment <- model_birds %>% 
  augment()

birds_augment
```

To run models or implement any function on the different groups within a data frame, one must master tidyr::nest() and purrr::map().

nest() creates a list of data frames with each corresponding to a different group.

Example: create a data frame for each taxa.

```{r}
data_taxa <- data_all %>% 
  group_by(Taxa) %>% 
  nest()

data_taxa
```

The grouping variable is left out of the nested data frame. All other data is now in a separate data frame, nested with the whole data frame.

You can inspect what lies behind the "data" column by using unnest() to display the original data.

```{r}
data_original_birds <- data_taxa %>% 
  ungroup() %>% 
  slice(n = 1) %>% 
  unnest(data) 

# slice(n = 1) means extract the first row of the data_taxa data frame

data_original_birds
```

map() applies a function to each nested data frame.

Example: find the mean number of species of each taxa

```{r}
mean_taxa <- data_all %>% 
  group_by(Taxa) %>% 
  nest() %>% 
  mutate(mean_richness = map(data, function(x) {
    mean(x %>% 
           pull(Richness))
    }
    )) 

# map indicates the data frame column with the nested data (i.e., data) and applies a function to that data.

# data becomes the "x" in the function. 

# pull() extracts the variable of interest and returns a vector.

mean_taxa
```

The results are mutated as another list column in the original data frame. They can be unnested.

```{r}
mean_taxa <- data_all %>% 
  group_by(Taxa) %>% 
  nest() %>% 
  mutate(mean_richness = map(data, function(x) {
    mean(x %>% 
           pull(Richness))
    }
    )) %>% 
  unnest(mean_richness)

mean_taxa
```

More complicated functions are easier to specify outside of the pipe.

Example: find the sites with the highest and lowest bird and butterfly species richness

```{r}
# Each element of the list column data will correspond to data frame "x" in the function.

get_sites <- function(x) { 
  x %>%
    filter(Richness %in% c(max(Richness), min(Richness))) %>% 
    mutate(Type = ifelse(Richness == max(Richness), "High", "Low")) %>% 
    arrange(desc(Richness))
}

# The function get.sites is applied to each element in the list column data.

sites_taxa <- data_all %>% 
  group_by(Taxa) %>% 
  nest() %>% 
  mutate(Types = map(data, get_sites)) %>% 
  unnest(Types) %>% 
  select(-data) 

sites_taxa
```

Another example, this time nesting the data based on sites and finding whether a site has more bird or butterfly species.

```{r}
get_taxa <- function(x) {
  Top_taxa<- x %>%
    filter(Richness == max(Richness)) %>%
    pull(Taxa)
 
 # Need to indicate sites where birds and butterflies have the same species richness 
   
  if(length(Top_taxa) > 1) {
    paste(Top_taxa[1], Top_taxa[2], sep=" + ")
    } else {
      Top_taxa
      } 
}

sites_richness <- data_all %>% 
  group_by(Site) %>% 
  nest() %>% 
  mutate(Animal = map(data, get_taxa)) %>%
  unnest(Animal) %>% 
  select(-data)

sites_richness
```

Are there any sites where birds and butterflies have equal species richness?

```{r}
site_animals <- sites_richness %>% 
  filter(Animal == "Birds + Butterflies")

site_animals
```

Are there more sites with higher bird or butterfly species richness?

```{r}
richness_summary <- sites_richness %>% 
  group_by(Animal) %>% 
  summarize(No_Sites = n())

richness_summary
```

Run a linear model to estimate the relationship between NDVI and richness for each taxa. 
"lm" can be specified as a function in map(). 
The model is mutated as a list column. We can mutate all model results by mapping tidy, glance, and augument onto each model.

```{r}
richness_models <- data_all %>% 
  group_by(Taxa) %>% 
  nest() %>% 
  mutate(Models = map(data, ~lm(Richness~NDVI, data=. ))) %>% 
  mutate(Tidied = map(Models, tidy), 
         Glanced = map(Models, glance), 
         Augmented = map(Models, augment))

richness_models
```

Is NDVI more strongly associated with bird or butterfly species richness? Let's compare the parameter estimates and their precision, which can be returned as a 95% confidence interval by specifying conf.int=TRUE and conf.level=0.95 (the default).

```{r}
richness_models <- data_all %>% 
  group_by(Taxa) %>% 
  nest() %>% 
  mutate(Models = map(data, ~lm(Richness~NDVI, data=. ))) %>% 
  mutate(Tidied = map(Models, tidy, conf.int = TRUE, conf.level = 0.95), 
         Glanced = map(Models, glance), 
         Augmented = map(Models, augment)) %>% 
  unnest(Tidied)

parameter_estimates <- richness_models %>% 
  select(Taxa, term, estimate, conf.low, conf.high) %>% 
  filter(term == "NDVI")

parameter_estimates
```

Graph the results: compare the relationship with NDVI for both species. Rather than use the fitted values, predict richness for all NDVI values from 0 to 1 at intervals of 0.01. The precision of the prediction can be estimated as confidence intervals (interval="confidence") or prediction intervals (interval="prediction")

```{r}
richness_models <- data_all %>% 
  group_by(Taxa) %>% 
  nest() %>% 
  mutate(Models = map(data, ~lm(Richness~NDVI, data=. ))) %>% 
  mutate(Tidied = map(Models, tidy), 
         Glanced = map(Models, glance),  
         Predictions = map(Models, augment, newdata = tibble(NDVI=seq(0,1,0.01)), 
                           se_fit=TRUE, interval="confidence")) 

# Rather than use the original NDVI values for prediction, 
# specify a new data frame (newdata) with the NDVI values of interest

richness_predictions <- richness_models %>% 
  unnest(Predictions) %>% 
  select(Taxa, NDVI, .fitted, .lower, .upper) %>% 
  rename(Richness = .fitted, low.95 = .lower, high.95 = .upper)

p <- richness_predictions %>% 
  ggplot(aes(x = NDVI, y = Richness, col = Taxa)) + 
  geom_line(size = 2) +
  
  # Visualize confidence interval
  
  geom_ribbon(aes(ymin = low.95, ymax = high.95, fill = Taxa), show.legend=FALSE, 
              color=NA, alpha=0.3) +
  
  # Add original data
  
  geom_point(data = data_all, aes(x = NDVI, y = Richness, col = Taxa), size=3) + 
  scale_y_continuous(name = "Species richness\n") + 
  scale_x_continuous(name = "\nNDVI") + 
  scale_color_viridis_d() + 
  scale_fill_viridis_d() + 
  theme_classic() + 
  theme(axis.line.x = element_line(colour = "black"), 
        axis.line.y = element_line(colour="black"), 
        text = element_text(size = 12))

p
#ggsave("Richness_vs_NDVI.pdf", device = "pdf", width = 6, height = 3, units = "in", dpi = 600)

```

