<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Session 7 Genetic and Environmental Associations | Reproducible Research for Conservation — Mérida, Mexico</title>
  <meta name="description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="generator" content="bookdown 0.24 and GitBook 2.6.7" />

  <meta property="og:title" content="Session 7 Genetic and Environmental Associations | Reproducible Research for Conservation — Mérida, Mexico" />
  <meta property="og:type" content="book" />
  
  
  <meta property="og:description" content="A book example for a Chapman &amp; Hall book." />
  <meta name="github-repo" content="yihui/bookdown-crc" />

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Session 7 Genetic and Environmental Associations | Reproducible Research for Conservation — Mérida, Mexico" />
  
  <meta name="twitter:description" content="A book example for a Chapman &amp; Hall book." />
  

<meta name="author" content="Eric C. Anderson, Christen Bossu, Richard Evan Feldman, Kristen C. Ruegg, Marius Somveille" />


<meta name="date" content="2022-01-12" />

  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="spatdata.html"/>
<link rel="next" href="conclude.html"/>
<script src="libs/header-attrs/header-attrs.js"></script>
<script src="libs/jquery/jquery-3.6.0.min.js"></script>
<script src="https://cdn.jsdelivr.net/npm/fuse.js@6.4.6/dist/fuse.min.js"></script>
<link href="libs/gitbook/css/style.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook/css/plugin-clipboard.css" rel="stylesheet" />








<link href="libs/anchor-sections/anchor-sections.css" rel="stylesheet" />
<script src="libs/anchor-sections/anchor-sections.js"></script>
<script src="libs/kePrint/kePrint.js"></script>
<link href="libs/lightable/lightable.css" rel="stylesheet" />


<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<style type="text/css">
/* Used with Pandoc 2.11+ new --citeproc when CSL is used */
div.csl-bib-body { }
div.csl-entry {
  clear: both;
}
.hanging div.csl-entry {
  margin-left:2em;
  text-indent:-2em;
}
div.csl-left-margin {
  min-width:2em;
  float:left;
}
div.csl-right-inline {
  margin-left:2em;
  padding-left:1em;
}
div.csl-indent {
  margin-left: 2em;
}
</style>

<link rel="stylesheet" href="css/style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">Mérida Workshop</a></li>

<li class="divider"></li>
<li class="chapter" data-level="" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i>Preface</a>
<ul>
<li class="chapter" data-level="0.1" data-path="index.html"><a href="index.html#setting-computer"><i class="fa fa-check"></i><b>0.1</b> Setting up your computer</a>
<ul>
<li class="chapter" data-level="0.1.1" data-path="index.html"><a href="index.html#step-1.-get-the-latest-versions-of-r-and-rstudio"><i class="fa fa-check"></i><b>0.1.1</b> Step 1. Get the latest versions of R and RStudio</a></li>
<li class="chapter" data-level="0.1.2" data-path="index.html"><a href="index.html#step-2.-install-a-number-of-r-packages-that-are-relatively-easy-to-install"><i class="fa fa-check"></i><b>0.1.2</b> Step 2. Install a number of R packages that are relatively easy to install</a></li>
<li class="chapter" data-level="0.1.3" data-path="index.html"><a href="index.html#step-3.-make-sure-you-have-git-and-an-account-on-github"><i class="fa fa-check"></i><b>0.1.3</b> Step 3. Make sure you have git and an account on GitHub</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="1" data-path="congen-intro.html"><a href="congen-intro.html"><i class="fa fa-check"></i><b>1</b> Introduction to Conservation Genomics Approaches</a>
<ul>
<li class="chapter" data-level="1.1" data-path="congen-intro.html"><a href="congen-intro.html#breakout-room-activity-material"><i class="fa fa-check"></i><b>1.1</b> Breakout room activity material</a></li>
<li class="chapter" data-level="1.2" data-path="congen-intro.html"><a href="congen-intro.html#after-kristens-talk-and-the-activity-please-download-the-course-repository"><i class="fa fa-check"></i><b>1.2</b> After Kristen’s Talk and the activity please download the course repository</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="basic-introduction-to-r-and-ggplot2.html"><a href="basic-introduction-to-r-and-ggplot2.html"><i class="fa fa-check"></i><b>2</b> Basic Introduction to R and <em>ggplot2</em></a>
<ul>
<li class="chapter" data-level="2.1" data-path="basic-introduction-to-r-and-ggplot2.html"><a href="basic-introduction-to-r-and-ggplot2.html#rstudio-layout"><i class="fa fa-check"></i><b>2.1</b> RStudio Layout</a></li>
<li class="chapter" data-level="2.2" data-path="basic-introduction-to-r-and-ggplot2.html"><a href="basic-introduction-to-r-and-ggplot2.html#data-visualization-with-ggplot2"><i class="fa fa-check"></i><b>2.2</b> Data visualization with ggplot2</a>
<ul>
<li class="chapter" data-level="2.2.1" data-path="basic-introduction-to-r-and-ggplot2.html"><a href="basic-introduction-to-r-and-ggplot2.html#understanding-the-ggplot-syntax"><i class="fa fa-check"></i><b>2.2.1</b> Understanding the Ggplot Syntax</a></li>
<li class="chapter" data-level="2.2.2" data-path="basic-introduction-to-r-and-ggplot2.html"><a href="basic-introduction-to-r-and-ggplot2.html#a-plot-consists-of-components"><i class="fa fa-check"></i><b>2.2.2</b> A plot consists of components</a></li>
</ul></li>
<li class="chapter" data-level="2.3" data-path="basic-introduction-to-r-and-ggplot2.html"><a href="basic-introduction-to-r-and-ggplot2.html#conclusion"><i class="fa fa-check"></i><b>2.3</b> Conclusion</a></li>
<li class="chapter" data-level="2.4" data-path="basic-introduction-to-r-and-ggplot2.html"><a href="basic-introduction-to-r-and-ggplot2.html#appendix-1-exercise-answers"><i class="fa fa-check"></i><b>2.4</b> Appendix 1: Exercise Answers</a></li>
</ul></li>
<li class="chapter" data-level="3" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html"><i class="fa fa-check"></i><b>3</b> Introduction to the tidyverse</a>
<ul>
<li class="chapter" data-level="3.1" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#what-tidyverse"><i class="fa fa-check"></i><b>3.1</b> What on earth is the tidyerse?</a></li>
<li class="chapter" data-level="3.2" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#an-example-data-set"><i class="fa fa-check"></i><b>3.2</b> An example data set</a></li>
<li class="chapter" data-level="3.3" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#reading-data-in-the-tidyverse"><i class="fa fa-check"></i><b>3.3</b> Reading data in the tidyverse</a></li>
<li class="chapter" data-level="3.4" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#a-word-on-tibbles"><i class="fa fa-check"></i><b>3.4</b> A word on tibbles</a></li>
<li class="chapter" data-level="3.5" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#back-to-tidying-up-our-genotype-data"><i class="fa fa-check"></i><b>3.5</b> Back to “tidying” up our genotype data</a></li>
<li class="chapter" data-level="3.6" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#what-the-heck-is-that-thingie"><i class="fa fa-check"></i><b>3.6</b> What the heck is that <code>%&gt;%</code> thingie?!</a></li>
<li class="chapter" data-level="3.7" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#the-dplyr-package-the-heart-of-the-tidyverse"><i class="fa fa-check"></i><b>3.7</b> The ‘dplyr’ package: the heart of the tidyverse</a>
<ul>
<li class="chapter" data-level="3.7.1" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#non-standard-evaluation-nse-you-dont-wrap-column-names-in-quotation-marks"><i class="fa fa-check"></i><b>3.7.1</b> Non-standard Evaluation (NSE): You don’t wrap column names in quotation marks</a></li>
<li class="chapter" data-level="3.7.2" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#select-to-select-columns-of-a-tibble"><i class="fa fa-check"></i><b>3.7.2</b> <code>select()</code> to select columns of a tibble</a></li>
<li class="chapter" data-level="3.7.3" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#filter-to-retain-certain-rows"><i class="fa fa-check"></i><b>3.7.3</b> <code>filter()</code> to retain certain rows</a></li>
<li class="chapter" data-level="3.7.4" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#arrange-to-reorder-the-rows-of-the-tibble"><i class="fa fa-check"></i><b>3.7.4</b> <code>arrange()</code> to reorder the rows of the tibble</a></li>
<li class="chapter" data-level="3.7.5" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#mutate-to-create-new-columns"><i class="fa fa-check"></i><b>3.7.5</b> <code>mutate()</code> to create new columns</a></li>
<li class="chapter" data-level="3.7.6" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#compute-a-single-value-from-many-rows-summarise"><i class="fa fa-check"></i><b>3.7.6</b> Compute a single value from many rows: <code>summarise()</code></a></li>
<li class="chapter" data-level="3.7.7" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#group_by-to-operate-tidily-on-subsets-of-a-tibble"><i class="fa fa-check"></i><b>3.7.7</b> <code>group_by()</code> to operate tidily on subsets of a tibble</a></li>
<li class="chapter" data-level="3.7.8" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#joins"><i class="fa fa-check"></i><b>3.7.8</b> Joins</a></li>
<li class="chapter" data-level="3.7.9" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#wrap-up"><i class="fa fa-check"></i><b>3.7.9</b> Wrap Up</a></li>
</ul></li>
<li class="chapter" data-level="3.8" data-path="introduction-to-the-tidyverse.html"><a href="introduction-to-the-tidyverse.html#modeling-within-the-tidyverse"><i class="fa fa-check"></i><b>3.8</b> Modeling within the Tidyverse</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="reporg.html"><a href="reporg.html"><i class="fa fa-check"></i><b>4</b> Organizing Research Projects for Reproducibility</a>
<ul>
<li class="chapter" data-level="4.1" data-path="reporg.html"><a href="reporg.html#goals-for-this-session"><i class="fa fa-check"></i><b>4.1</b> Goals for this session</a></li>
<li class="chapter" data-level="4.2" data-path="reporg.html"><a href="reporg.html#why-do-this"><i class="fa fa-check"></i><b>4.2</b> Why do this?</a></li>
<li class="chapter" data-level="4.3" data-path="reporg.html"><a href="reporg.html#what-can-this-look-like"><i class="fa fa-check"></i><b>4.3</b> What can this look like?</a></li>
<li class="chapter" data-level="4.4" data-path="reporg.html"><a href="reporg.html#reproducibility-edict-1-dont-save-your-r-workspace-and-then-reload-it"><i class="fa fa-check"></i><b>4.4</b> Reproducibility Edict #1: Don’t Save Your R Workspace and then Reload It</a></li>
<li class="chapter" data-level="4.5" data-path="reporg.html"><a href="reporg.html#paths-an-important-concept-for-reproducibility"><i class="fa fa-check"></i><b>4.5</b> Paths, an important concept for reproducibility</a></li>
<li class="chapter" data-level="4.6" data-path="reporg.html"><a href="reporg.html#the-working-directory"><i class="fa fa-check"></i><b>4.6</b> The Working Directory</a>
<ul>
<li class="chapter" data-level="4.6.1" data-path="reporg.html"><a href="reporg.html#a-hugely-important-point"><i class="fa fa-check"></i><b>4.6.1</b> A Hugely Important Point!</a></li>
<li class="chapter" data-level="4.6.2" data-path="reporg.html"><a href="reporg.html#skeletal-anatomy-of-an-rstudio-project"><i class="fa fa-check"></i><b>4.6.2</b> Skeletal anatomy of an RStudio Project</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="reporg.html"><a href="reporg.html#reproducibility-edict-2-dont-change-directories-in-your-code"><i class="fa fa-check"></i><b>4.7</b> Reproducibility Edict #2: Don’t Change Directories in Your Code</a></li>
<li class="chapter" data-level="4.8" data-path="reporg.html"><a href="reporg.html#relative-paths"><i class="fa fa-check"></i><b>4.8</b> Relative Paths</a></li>
<li class="chapter" data-level="4.9" data-path="reporg.html"><a href="reporg.html#reproducibility-edict-3-always-use-relative-paths-from-the-project-directory-to-files-inside-the-project-directory"><i class="fa fa-check"></i><b>4.9</b> Reproducibility Edict #3: Always Use Relative Paths from the Project Directory to Files <em>inside</em> the Project Directory</a></li>
<li class="chapter" data-level="4.10" data-path="reporg.html"><a href="reporg.html#reproducibility-edict-4-put-data-files-into-a-data-directory"><i class="fa fa-check"></i><b>4.10</b> Reproducibility Edict #4: Put data files into a <code>data</code> directory</a></li>
<li class="chapter" data-level="4.11" data-path="reporg.html"><a href="reporg.html#reproducibility-edict-5-write-analysis-scripts-as-.r-or-.rmd-files"><i class="fa fa-check"></i><b>4.11</b> Reproducibility Edict #5: Write Analysis Scripts as <code>.R</code> or <code>.Rmd</code> files</a></li>
<li class="chapter" data-level="4.12" data-path="reporg.html"><a href="reporg.html#download-an-example-project-to-play-with"><i class="fa fa-check"></i><b>4.12</b> Download an example project to play with</a>
<ul>
<li class="chapter" data-level="4.12.1" data-path="reporg.html"><a href="reporg.html#lets-have-a-moment"><i class="fa fa-check"></i><b>4.12.1</b> Let’s have a moment</a></li>
</ul></li>
<li class="chapter" data-level="4.13" data-path="reporg.html"><a href="reporg.html#reproducibility-edict-6-write-your-code-with-style"><i class="fa fa-check"></i><b>4.13</b> Reproducibility Edict #6: Write your code with style!</a></li>
<li class="chapter" data-level="4.14" data-path="reporg.html"><a href="reporg.html#what-about-shared-data-sets"><i class="fa fa-check"></i><b>4.14</b> What About Shared Data Sets?</a></li>
<li class="chapter" data-level="4.15" data-path="reporg.html"><a href="reporg.html#reproducibility-edict-7-use-git-for-version-control"><i class="fa fa-check"></i><b>4.15</b> Reproducibility Edict #7: Use git for Version Control</a></li>
<li class="chapter" data-level="4.16" data-path="reporg.html"><a href="reporg.html#in-the-meantime-a-short-demo-of-gitgithub-on-rstudio"><i class="fa fa-check"></i><b>4.16</b> In the meantime, a short demo of git/GitHub on RStudio</a></li>
<li class="chapter" data-level="4.17" data-path="reporg.html"><a href="reporg.html#what-to-keep-under-version-control-and-what-not-to"><i class="fa fa-check"></i><b>4.17</b> What to keep under version control and what not to</a></li>
<li class="chapter" data-level="4.18" data-path="reporg.html"><a href="reporg.html#create-a-readme-to-describe-the-parts-of-your-analysis-project"><i class="fa fa-check"></i><b>4.18</b> Create a README to describe the parts of your analysis project</a></li>
<li class="chapter" data-level="4.19" data-path="reporg.html"><a href="reporg.html#github-and-very-large-data-sets"><i class="fa fa-check"></i><b>4.19</b> GitHub and Very large data sets</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="envdata.html"><a href="envdata.html"><i class="fa fa-check"></i><b>5</b> Finding and Obtaining Environmental Data for Conservation</a>
<ul>
<li class="chapter" data-level="5.1" data-path="envdata.html"><a href="envdata.html#environmental-layers.-a-general-overview"><i class="fa fa-check"></i><b>5.1</b> Environmental Layers. A general overview</a>
<ul>
<li class="chapter" data-level="5.1.1" data-path="envdata.html"><a href="envdata.html#environmental-layers-in-conservation-genetics"><i class="fa fa-check"></i><b>5.1.1</b> Environmental layers in conservation genetics</a></li>
<li class="chapter" data-level="5.1.2" data-path="envdata.html"><a href="envdata.html#examples-of-atmospheric-layers"><i class="fa fa-check"></i><b>5.1.2</b> Examples of atmospheric layers</a></li>
<li class="chapter" data-level="5.1.3" data-path="envdata.html"><a href="envdata.html#examples-of-vegetation-layers"><i class="fa fa-check"></i><b>5.1.3</b> Examples of vegetation layers</a></li>
<li class="chapter" data-level="5.1.4" data-path="envdata.html"><a href="envdata.html#examples-of-landscape-layers"><i class="fa fa-check"></i><b>5.1.4</b> Examples of landscape layers</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="6" data-path="spatdata.html"><a href="spatdata.html"><i class="fa fa-check"></i><b>6</b> Handling and Visualizing Spatial Data in R</a>
<ul>
<li class="chapter" data-level="6.1" data-path="spatdata.html"><a href="spatdata.html#handling-spatial-rasters-in-r"><i class="fa fa-check"></i><b>6.1</b> Handling Spatial Rasters in R</a>
<ul>
<li class="chapter" data-level="6.1.1" data-path="spatdata.html"><a href="spatdata.html#reading-and-printing-raster-data"><i class="fa fa-check"></i><b>6.1.1</b> Reading and Printing Raster Data</a></li>
<li class="chapter" data-level="6.1.2" data-path="spatdata.html"><a href="spatdata.html#plotting-raster-data-with-plot"><i class="fa fa-check"></i><b>6.1.2</b> Plotting Raster Data with <code>plot()</code></a></li>
<li class="chapter" data-level="6.1.3" data-path="spatdata.html"><a href="spatdata.html#plotting-raster-data-with-ggplot2"><i class="fa fa-check"></i><b>6.1.3</b> Plotting raster data with ggplot2</a></li>
</ul></li>
<li class="chapter" data-level="6.2" data-path="spatdata.html"><a href="spatdata.html#vector-spatial-data"><i class="fa fa-check"></i><b>6.2</b> Vector Spatial Data</a>
<ul>
<li class="chapter" data-level="6.2.1" data-path="spatdata.html"><a href="spatdata.html#vector-data-features"><i class="fa fa-check"></i><b>6.2.1</b> Vector data features</a></li>
<li class="chapter" data-level="6.2.2" data-path="spatdata.html"><a href="spatdata.html#shapefiles"><i class="fa fa-check"></i><b>6.2.2</b> Shapefiles</a></li>
<li class="chapter" data-level="6.2.3" data-path="spatdata.html"><a href="spatdata.html#spatial-vector-data-in-r"><i class="fa fa-check"></i><b>6.2.3</b> Spatial vector data in R</a></li>
<li class="chapter" data-level="6.2.4" data-path="spatdata.html"><a href="spatdata.html#an-example-mexican-states"><i class="fa fa-check"></i><b>6.2.4</b> An example: Mexican states</a></li>
<li class="chapter" data-level="6.2.5" data-path="spatdata.html"><a href="spatdata.html#plotting-simple-features-with-ggplot"><i class="fa fa-check"></i><b>6.2.5</b> Plotting simple features with <code>ggplot()</code></a></li>
<li class="chapter" data-level="6.2.6" data-path="spatdata.html"><a href="spatdata.html#a-word-about-coord_sf"><i class="fa fa-check"></i><b>6.2.6</b> A word about <code>coord_sf()</code></a></li>
<li class="chapter" data-level="6.2.7" data-path="spatdata.html"><a href="spatdata.html#a-word-about-resolution-of-spatial-vector-data"><i class="fa fa-check"></i><b>6.2.7</b> A word about <em>resolution</em> of spatial vector data</a></li>
</ul></li>
<li class="chapter" data-level="6.3" data-path="spatdata.html"><a href="spatdata.html#yucatan-jay-occurrences"><i class="fa fa-check"></i><b>6.3</b> Yucatan Jay Occurrences</a>
<ul>
<li class="chapter" data-level="6.3.1" data-path="spatdata.html"><a href="spatdata.html#example-data-set-with-latitude-and-longitude-points"><i class="fa fa-check"></i><b>6.3.1</b> Example data set with Latitude and Longitude points</a></li>
<li class="chapter" data-level="6.3.2" data-path="spatdata.html"><a href="spatdata.html#plotting-lat-long-data"><i class="fa fa-check"></i><b>6.3.2</b> Plotting Lat-Long data</a></li>
<li class="chapter" data-level="6.3.3" data-path="spatdata.html"><a href="spatdata.html#but-geom_point-wont-work-with-different-projections-coordinate-reference-sysytems"><i class="fa fa-check"></i><b>6.3.3</b> But, <code>geom_point()</code> won’t work with Different Projections / Coordinate Reference Sysytems</a></li>
<li class="chapter" data-level="6.3.4" data-path="spatdata.html"><a href="spatdata.html#turning-something-into-a-simple-features-object"><i class="fa fa-check"></i><b>6.3.4</b> Turning something into a simple features object</a></li>
</ul></li>
<li class="chapter" data-level="6.4" data-path="spatdata.html"><a href="spatdata.html#geometric-operations"><i class="fa fa-check"></i><b>6.4</b> Geometric operations</a>
<ul>
<li class="chapter" data-level="6.4.1" data-path="spatdata.html"><a href="spatdata.html#spatial-intersection"><i class="fa fa-check"></i><b>6.4.1</b> Spatial Intersection</a></li>
<li class="chapter" data-level="6.4.2" data-path="spatdata.html"><a href="spatdata.html#treat-the-result-as-a-tibble"><i class="fa fa-check"></i><b>6.4.2</b> Treat the result as a tibble</a></li>
<li class="chapter" data-level="6.4.3" data-path="spatdata.html"><a href="spatdata.html#the-same-filtering-can-be-done-on-the-state-polygons"><i class="fa fa-check"></i><b>6.4.3</b> The same filtering can be done on the state polygons</a></li>
</ul></li>
<li class="chapter" data-level="6.5" data-path="spatdata.html"><a href="spatdata.html#joint-operations-with-raster-and-vector-data-i-masking-a-raster-with-a-polygon"><i class="fa fa-check"></i><b>6.5</b> Joint Operations with Raster and Vector data I: Masking a Raster with a Polygon</a></li>
<li class="chapter" data-level="6.6" data-path="spatdata.html"><a href="spatdata.html#joint-operations-with-raster-and-vector-data-ii-extracting-raster-values-at-a-set-of-spatial-points"><i class="fa fa-check"></i><b>6.6</b> Joint Operations with Raster and Vector data II: Extracting raster values at a set of spatial points</a></li>
<li class="chapter" data-level="6.7" data-path="spatdata.html"><a href="spatdata.html#spat_data_appendix"><i class="fa fa-check"></i><b>6.7</b> Appendix: Obtaining and Reducing The Example Spatial Datasets</a>
<ul>
<li class="chapter" data-level="6.7.1" data-path="spatdata.html"><a href="spatdata.html#raster-data"><i class="fa fa-check"></i><b>6.7.1</b> Raster Data</a></li>
<li class="chapter" data-level="6.7.2" data-path="spatdata.html"><a href="spatdata.html#vector-data"><i class="fa fa-check"></i><b>6.7.2</b> Vector data</a></li>
<li class="chapter" data-level="6.7.3" data-path="spatdata.html"><a href="spatdata.html#species-occurrence-data"><i class="fa fa-check"></i><b>6.7.3</b> Species Occurrence Data</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="7" data-path="geassoc.html"><a href="geassoc.html"><i class="fa fa-check"></i><b>7</b> Genetic and Environmental Associations</a>
<ul>
<li class="chapter" data-level="7.1" data-path="geassoc.html"><a href="geassoc.html#background"><i class="fa fa-check"></i><b>7.1</b> Background</a></li>
<li class="chapter" data-level="7.2" data-path="geassoc.html"><a href="geassoc.html#motivation"><i class="fa fa-check"></i><b>7.2</b> Motivation</a></li>
<li class="chapter" data-level="7.3" data-path="geassoc.html"><a href="geassoc.html#housekeeping"><i class="fa fa-check"></i><b>7.3</b> Housekeeping</a></li>
<li class="chapter" data-level="7.4" data-path="geassoc.html"><a href="geassoc.html#running-gradient-forest"><i class="fa fa-check"></i><b>7.4</b> Running Gradient Forest</a></li>
<li class="chapter" data-level="7.5" data-path="geassoc.html"><a href="geassoc.html#basic-gradient-forest-preparation"><i class="fa fa-check"></i><b>7.5</b> Basic Gradient Forest Preparation</a></li>
<li class="chapter" data-level="7.6" data-path="geassoc.html"><a href="geassoc.html#defining-genomic-vulnerabilitygenomic-offset-given-future-climate-predictions"><i class="fa fa-check"></i><b>7.6</b> Defining “Genomic Vulnerability/Genomic offset” Given Future Climate Predictions</a></li>
<li class="chapter" data-level="7.7" data-path="geassoc.html"><a href="geassoc.html#next-steps"><i class="fa fa-check"></i><b>7.7</b> Next steps</a>
<ul>
<li class="chapter" data-level="7.7.1" data-path="geassoc.html"><a href="geassoc.html#randomization"><i class="fa fa-check"></i><b>7.7.1</b> Randomization</a></li>
<li class="chapter" data-level="7.7.2" data-path="geassoc.html"><a href="geassoc.html#candidate-variants"><i class="fa fa-check"></i><b>7.7.2</b> Candidate variants</a></li>
</ul></li>
</ul></li>
<li class="chapter" data-level="8" data-path="conclude.html"><a href="conclude.html"><i class="fa fa-check"></i><b>8</b> Concluding Remarks / Parting Words</a></li>
<li class="divider"></li>
<li><a href="https://bookdown.org" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./">Reproducible Research for Conservation — Mérida, Mexico</a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="geassoc" class="section level1" number="7">
<h1><span class="header-section-number">Session 7</span> Genetic and Environmental Associations</h1>
<div id="background" class="section level2" number="7.1">
<h2><span class="header-section-number">7.1</span> Background</h2>
<p>Now that we know how to get environmental layers and plot them in space, for many of the species we are working on, we are not entirely sure which environmental variables are important drivers of populaiton genetic structure or underlying phenotypic traits of interest.</p>
<p>Let’s back up. Now, we have access to multiple layers of environmental layers that span a species range:</p>
<p><img src="images/env_layers/three-layers-intro.png" width="100%" /></p>
<p>And many of us now have genomic data. Here, we used genome-wide reduced representation sequencing to create a spatially explicit map of population structure, called the genoscape, of the Willow Flycatcher.</p>
<p><img src="images/6.0/Ruegg%20et%20al2021.Fig1.jpg" width="90%" /></p>
<p>What we ultimately want to investigate is how environmental variables are correlated with genetic variation.</p>
<p><img src="images/6.0/GEA.Fig1.jpeg" width="90%" /></p>
<p>Standing genetic variation associated with current climate gradients theoretically indicates adaptive capacity. However, modeling future adaptation from individual loci is difficult because we don’t understand all GxE interactions.</p>
<p>A recent example illustrated that not only is geography (“isolation by distance”) strongly correlated genetic structure, but environment plays a role in differentiation too. Specifically, precipitation is a strong predictor of Yellow Warbler genetic structure.</p>
<p><img src="images/6.0/Bay%20et%20al.Fig1.jpg" width="100%" /></p>
</div>
<div id="motivation" class="section level2" number="7.2">
<h2><span class="header-section-number">7.2</span> Motivation</h2>
<p>The R package gradientForest traditionally quantifies multi-species compositional turnover along environmental gradients. In the era of next-generation sequencing, where genome-wide data is common, we can extend this community ecology framework to determine which variants are associated with specific environmental variables. So rather than thinking of the large-scale species turnover in space, we can map fine-scale allele frequency changes along environemntal gradients.</p>
<p>Gradient forest is a machine learning approach for classification and regression based on decision trees. We’ve all seen and used decision trees, for mundane tasks like waking up to an alarm each morning to more intricate decisions like optimal pairing of individuals in a captive breeding program to increase/maintain genetic diversity of a population. Regardless of which decision tree you are implementing, the assumption remains, the higher the split, the more important the predictor.</p>
<p>Because decision trees are prone to overiftting, gradient forest uses random forests to test associations. Random forests use a random subset of predictors for each decision tree.</p>
<p><img src="images/6.0/RandomForest.ConceptMap.jpg" width="100%" /></p>
<ol style="list-style-type: decimal">
<li>Choose a subset of data and predictors.</li>
<li>Find predictor (and value) that optimizes split.</li>
<li>Create many trees and aggregate.</li>
</ol>
<p><strong>Random forests and genomic data</strong></p>
<p>Brieuc et al. 2018 used this approach to rank the importance of SNPs for predicting a specific binary phenotype, <em>e.g.</em> disease resistance in fish. This is an example of one random tree, which can ultimately create a predictive polygenic model.</p>
<p><img src="images/6.0/RandomForest.Genomics.jpg" width="100%" /></p>
<p>Another interesting study used random forests to find the predictive variants for cold hardiness, which was subsequently used to predict cold injury in the Sitka Spruce (Holliday et al. 2012).</p>
<p><img src="images/6.0/RandomForest.Genomics2.jpg" width="100%" /></p>
<p><strong>What if we have multiple SNPs AND multiple phenotypes or environments?</strong>
Gradient forest extends random forest to incorporate multiple response variables.</p>
<p>Gradient uses turnover functions to look at total allele turnover across environmental gradients. For instance, GI5, a candidate gene in Balsam Poplar, clearly shows great allele frequency transitions across the mean summer temperature (Fitzpatrick and Keller 2015)</p>
<p><img src="images/6.0/AlleleTurnover.jpg" width="100%" /></p>
<p>If we look at different subsets of genetic variation in Balsam Poplar space</p>
<ol style="list-style-type: lower-alpha">
<li>reference SNPs,<br />
</li>
<li>GI5, a candidate gene for genetic control of bud set and bud flush (Olson et al. 2013)</li>
</ol>
<p>you can see that different environmental variables predict genetic variation of the reference and candidate genes. <em>Note</em> different colors = locally adapted genotypes.</p>
<p><img src="images/6.0/Fitzpatrick.Fig5.jpeg" width="100%" /></p>
<p>Below is code for quantifying and visualizing intraspecific gene-environment variation across space. As part of The Bird Genoscape Project, we compare present-day gene-environment associations to predicted future changes to identify regions of genomic vulnerability.</p>
<p><img src="images/6.0/genomic_vulnerability.jpg" width="60%" /></p>
<p>For now, this tutorial focuses on analyzing and plotting the allelic variation as estimated from gradient forest for a single species, the Willow Flycatcher, but is easily adpated to include other species and methods for esimating variation.</p>
</div>
<div id="housekeeping" class="section level2" number="7.3">
<h2><span class="header-section-number">7.3</span> Housekeeping</h2>
<p>We will need the following R packages. gradientForest requires the dependancy extendedForest, an R package classification and regression based on forest trees using random inputs. With the newest version of R, extended forest is installed automatically as a dependency.</p>
<p>Install and load the following packages in Rstudio. You can use the Packages function to the right to install most packages except for gradient forest, which we will install from R-Forge.</p>
<ol style="list-style-type: decimal">
<li>As discussed in <a href="index.html#setting-computer">Setting Up Your Computer</a>, if you have a Mac, you
will need the XCode command line tools.</li>
<li>If you have a PC, you’ll need Rtools</li>
<li>You will also need gfortran on your computer. See the instructions in
<a href="index.html#setting-computer">Setting Up Your Computer</a> if you don’t have that set up
already.</li>
</ol>
<p>Also, you need to have the package ‘gradientForest’ installed as
described in the Preface section <a href="index.html#grad-forest">The ‘gradientForest’ package</a>.</p>
<div class="sourceCode" id="cb278"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb278-1"><a href="geassoc.html#cb278-1" aria-hidden="true" tabindex="-1"></a><span class="co">#install.packages(&quot;gradientForest&quot;, repos=&quot;http://R-Forge.R-project.org&quot;)</span></span>
<span id="cb278-2"><a href="geassoc.html#cb278-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gradientForest)</span>
<span id="cb278-3"><a href="geassoc.html#cb278-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(data.table)</span></code></pre></div>
<pre><code>## 
## Attaching package: &#39;data.table&#39;</code></pre>
<pre><code>## The following objects are masked from &#39;package:terra&#39;:
## 
##     copy, shift</code></pre>
<pre><code>## The following objects are masked from &#39;package:dplyr&#39;:
## 
##     between, first, last</code></pre>
<pre><code>## The following object is masked from &#39;package:purrr&#39;:
## 
##     transpose</code></pre>
<div class="sourceCode" id="cb283"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb283-1"><a href="geassoc.html#cb283-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(raster)</span></code></pre></div>
<pre><code>## Loading required package: sp</code></pre>
<pre><code>## 
## Attaching package: &#39;raster&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:dplyr&#39;:
## 
##     select</code></pre>
<div class="sourceCode" id="cb287"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb287-1"><a href="geassoc.html#cb287-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(gstat)</span>
<span id="cb287-2"><a href="geassoc.html#cb287-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rgdal)</span></code></pre></div>
<pre><code>## Please note that rgdal will be retired by the end of 2023,
## plan transition to sf/stars/terra functions using GDAL and PROJ
## at your earliest convenience.
## 
## rgdal: version: 1.5-28, (SVN revision 1158)
## Geospatial Data Abstraction Library extensions to R successfully loaded
## Loaded GDAL runtime: GDAL 3.2.1, released 2020/12/29
## Path to GDAL shared files: /Library/Frameworks/R.framework/Versions/4.1/Resources/library/rgdal/gdal
## GDAL binary built with GEOS: TRUE 
## Loaded PROJ runtime: Rel. 7.2.1, January 1st, 2021, [PJ_VERSION: 721]
## Path to PROJ shared files: /Library/Frameworks/R.framework/Versions/4.1/Resources/library/rgdal/proj
## PROJ CDN enabled: FALSE
## Linking to sp version:1.4-6
## To mute warnings of possible GDAL/OSR exportToProj4() degradation,
## use options(&quot;rgdal_show_exportToProj4_warnings&quot;=&quot;none&quot;) before loading sp or rgdal.
## Overwritten PROJ_LIB was /Library/Frameworks/R.framework/Versions/4.1/Resources/library/rgdal/proj</code></pre>
<pre><code>## 
## Attaching package: &#39;rgdal&#39;</code></pre>
<pre><code>## The following object is masked from &#39;package:terra&#39;:
## 
##     project</code></pre>
<div class="sourceCode" id="cb291"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb291-1"><a href="geassoc.html#cb291-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sf)</span>
<span id="cb291-2"><a href="geassoc.html#cb291-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(RColorBrewer)</span>
<span id="cb291-3"><a href="geassoc.html#cb291-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(rasterVis)</span></code></pre></div>
<pre><code>## Loading required package: lattice</code></pre>
<div class="sourceCode" id="cb293"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb293-1"><a href="geassoc.html#cb293-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span></code></pre></div>
<p>And we will need to load a realistic, polygon shapefile (an outline of the Willow Flycatcher’s breeding range in North America for this exercise) and set its projection. <strong>Note, you will need to adjust the path to the shapefile below. This is the path to my shape file on my computer.</strong></p>
<p>That will produce this map:</p>
<div class="sourceCode" id="cb294"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb294-1"><a href="geassoc.html#cb294-1" aria-hidden="true" tabindex="-1"></a>wifl_shape <span class="ot">&lt;-</span><span class="fu">st_read</span>(<span class="st">&quot;data/6.0/WIFLrev.shp&quot;</span>)</span></code></pre></div>
<pre><code>## Reading layer `WIFLrev&#39; from data source 
##   `/Users/eriq/Documents/git-repos/merida-workshop-2022/data/6.0/WIFLrev.shp&#39; 
##   using driver `ESRI Shapefile&#39;
## Simple feature collection with 5 features and 14 fields
## Geometry type: MULTIPOLYGON
## Dimension:     XY
## Bounding box:  xmin: -128.4 ymin: 29.04 xmax: -67.01 ymax: 53.46
## Geodetic CRS:  WGS 84</code></pre>
<div class="sourceCode" id="cb296"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb296-1"><a href="geassoc.html#cb296-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggplot</span>(wifl_shape) <span class="sc">+</span></span>
<span id="cb296-2"><a href="geassoc.html#cb296-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">fill=</span><span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb296-3"><a href="geassoc.html#cb296-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>()</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:unnamed-chunk-185"></span>
<img src="merida-workshop-2022_files/figure-html/unnamed-chunk-185-1.png" alt="A basic outline of Willow Flycatcher breeding range we will use for this exercise" width="672" />
<p class="caption">
FIGURE 7.1: A basic outline of Willow Flycatcher breeding range we will use for this exercise
</p>
</div>
</div>
<div id="running-gradient-forest" class="section level2" number="7.4">
<h2><span class="header-section-number">7.4</span> Running Gradient Forest</h2>
<p>We will use gradient forest to test whether a subset of genomic variation can be explained by environmental variables and to visualize climate-associated genetic variation across the breeding range of the Willow Flycatcher. To run gradient forest we need two main files. First we need the environmental data/climate variables associated with each population. Populations are defined by their latitude and longitude.</p>
<p>For each sampling location, we obtained environmental data from publically available databases. Specifically, we chose variables with known impacts on bird physiology and ecology and that were publically available at high resolution. These variables included 19 climate variables downloaded from WorldClim (Hijmans et al. 2005) as well as vegetation indices (Carroll et al. 2004) (NDVI and NDVIstd), Tree Cover (Sexton et al. 2013) and elevation data from the Global Land Cover Facility (<a href="http://www.landcover.org" class="uri">http://www.landcover.org</a>). We also downloaded and used a measure of surface moisture characteristics from the NASA Scatterometer Climate Record Pathfinder project (QuickSCAT, downloaded from scp.byu.edu).</p>
<p>Read in the environmental variables, and take a look at the dimensions and the actual data.</p>
<div class="sourceCode" id="cb297"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb297-1"><a href="geassoc.html#cb297-1" aria-hidden="true" tabindex="-1"></a>Ewifl<span class="ot">&lt;-</span><span class="fu">read_csv</span>(<span class="st">&#39;data/6.0/wiflforest.env.csv&#39;</span>)</span></code></pre></div>
<pre><code>## Rows: 21 Columns: 23</code></pre>
<pre><code>## ── Column specification ───────────────────────────────
## Delimiter: &quot;,&quot;
## dbl (23): BIO1, BIO2, BIO3, BIO4, BIO5, BIO6, BIO7,...</code></pre>
<pre><code>## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div class="sourceCode" id="cb301"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb301-1"><a href="geassoc.html#cb301-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(Ewifl)</span></code></pre></div>
<pre><code>## [1] 21 23</code></pre>
<p><strong>How many sites do we have environmental data for? How many environmental variables?</strong></p>
<p>Gradient forest does not directly account for spatial population structure. But there are some possible ways to deal with this:</p>
<ul>
<li><p>Include Lat/Long as predictors.</p></li>
<li><p>Moran’s Eigenvector Map (MEM) variables – spatial eigenvectors (Fitzpatrick et al. 2015)</p></li>
<li><p>”regress out” population cluster membership (Holliday et al. 2012 did this for phenotypes).</p></li>
<li><p>Use genomic PCAs as predictors (Exposito-Alonso et al. 2018)</p></li>
</ul>
<p>Your choice depends on the biology of your organism and your question. You’ll notice in this tutorial, by examining the environmental data we are using in the code chunk below, we do not account for population structure.</p>
<div class="sourceCode" id="cb303"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb303-1"><a href="geassoc.html#cb303-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(Ewifl)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 23
##    BIO1  BIO2  BIO3  BIO4  BIO5  BIO6  BIO7  BIO8  BIO9
##   &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;
## 1   193   165    43  7888   392    15   377    95   227
## 2   205   165    44  7792   395    23   372   300   242
## 3    52   180    50  6365   237  -116   353   131    78
## 4    86   166    47  6468   290   -63   353    13   171
## 5   168   107    54  3298   265    65   200   132   206
## 6   166   158    44  7032   360     4   356    94   258
## # … with 14 more variables: BIO10 &lt;dbl&gt;, BIO11 &lt;dbl&gt;,
## #   BIO12 &lt;dbl&gt;, BIO13 &lt;dbl&gt;, BIO14 &lt;dbl&gt;,
## #   BIO15 &lt;dbl&gt;, BIO16 &lt;dbl&gt;, BIO17 &lt;dbl&gt;,
## #   BIO18 &lt;dbl&gt;, BIO19 &lt;dbl&gt;, NDVI_Mean &lt;dbl&gt;,
## #   NDVI_StDev &lt;dbl&gt;, SRTM &lt;dbl&gt;, TreeCover &lt;dbl&gt;</code></pre>
<p>Second, we have genomic data formatted as allele frequencies for each population in the same order. The published Willow Flycatcher data includes 37855 variants, and is too large to run a gradient forest analyses within the time frame of a tutorial, so we are supplying a subsetted dataset limited to 10,000 SNPs.</p>
<div class="sourceCode" id="cb305"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb305-1"><a href="geassoc.html#cb305-1" aria-hidden="true" tabindex="-1"></a>Gwifl<span class="ot">&lt;-</span><span class="fu">read_csv</span>(<span class="st">&#39;data/6.0/wiflforest.allfreq.sample.csv&#39;</span>)</span></code></pre></div>
<pre><code>## Rows: 21 Columns: 10000</code></pre>
<pre><code>## ── Column specification ───────────────────────────────
## Delimiter: &quot;,&quot;
## dbl (10000): V21252, V9139, V22510, V5860, V3386, V...</code></pre>
<pre><code>## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div class="sourceCode" id="cb309"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb309-1"><a href="geassoc.html#cb309-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(Gwifl)</span></code></pre></div>
<pre><code>## [1]    21 10000</code></pre>
<div class="sourceCode" id="cb311"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb311-1"><a href="geassoc.html#cb311-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(Gwifl)</span></code></pre></div>
<pre><code>## # A tibble: 6 × 10,000
##   V21252  V9139 V22510 V5860 V3386 V14932 V14581  V9277
##    &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;  &lt;dbl&gt;
## 1  0.15  0.333  0.25   0.45  0.3    0.3    0.4   0.125 
## 2  0.107 0.25   0.25   0.571 0.308  0.607  0.357 0.143 
## 3  0.115 0.231  0.167  0.654 0.308  0.208  0.192 0.136 
## 4  0     0.1    0.0833 0.667 0.417  0.167  0     0.1   
## 5  0.25  0.667  0.125  0.625 0      0.75   0.25  0     
## 6  0.192 0.0833 0.0385 0.115 0.192  0.269  0.154 0.0909
## # … with 9,992 more variables: V25872 &lt;dbl&gt;,
## #   V23716 &lt;dbl&gt;, V25026 &lt;dbl&gt;, V11453 &lt;dbl&gt;,
## #   V2266 &lt;dbl&gt;, V371 &lt;dbl&gt;, V21324 &lt;dbl&gt;,
## #   V28422 &lt;dbl&gt;, V8773 &lt;dbl&gt;, V5906 &lt;dbl&gt;,
## #   V15909 &lt;dbl&gt;, V20940 &lt;dbl&gt;, V15435 &lt;dbl&gt;,
## #   V9026 &lt;dbl&gt;, V12438 &lt;dbl&gt;, V2181 &lt;dbl&gt;,
## #   V23539 &lt;dbl&gt;, V19510 &lt;dbl&gt;, V15571 &lt;dbl&gt;, …</code></pre>
<p><strong>These are allele frequencies for each site. What is the range of allele frequencies for allele V3386? You might need to write extra code to answer this question.</strong></p>
<p>Next we have some housekeeping to define the predictor variables in the regression, which are the environmental variables, and the response variables, which in this case are the frequencies of each variant, analyzed in gradient forest.</p>
<p>Another variable in the <em>gradientForest</em> model is <em>maxLevel</em>, which we define using an equation recommended by gradient forest.</p>
<div class="sourceCode" id="cb313"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb313-1"><a href="geassoc.html#cb313-1" aria-hidden="true" tabindex="-1"></a><span class="co"># set depth of conditional permutation</span></span>
<span id="cb313-2"><a href="geassoc.html#cb313-2" aria-hidden="true" tabindex="-1"></a>lev <span class="ot">&lt;-</span> <span class="fu">floor</span>(<span class="fu">log2</span>(nSites<span class="sc">*</span><span class="fl">0.368</span><span class="sc">/</span><span class="dv">2</span>))</span>
<span id="cb313-3"><a href="geassoc.html#cb313-3" aria-hidden="true" tabindex="-1"></a>lev</span></code></pre></div>
<pre><code>## [1] 1</code></pre>
<p>First, and notably, predictor variables are often correlated. The standard approach in random forests assesses marginal importance of predictor
by randomly permuting each predictor in turn, across all sites in the dataset,
and calculating the degradation prediction performance of each tree. Package
<em>extendedForest</em> can account for correlated predictors by implementing
conditional permutation <strong>Ellis2010</strong>, following the strategy outlined
by <strong>Strobl2008</strong>. In conditional permutation, the predictor to be
assessed is permuted only within blocks of the dataset defined by splits in the
given tree on any other predictors correlated above a certain threshold (e.g. 
r= 0.5) and up to a maximum number of splits set by the <em>maxLevel</em> option. A <strong>maxLevel &gt; 0</strong> will compute importance from conditional permutation distribution of each variable, permuted within <em>2^maxLevel</em> partitions of correlated variables.</p>
<p><em>Note:</em> the <em>ntree</em> default is 10, however, more is better and reduces variance. Gradient forest recommends 500 for a true run. Also, gradient forest bins variants, thus large scale patterns will be seen, however, if there is one variant of large effect, it may not be found in this analysis.</p>
<p>If you are asking about ecological drivers of a small number of variants underlying a trait of interest (candidate variants identified using GWAS), then you wouldn’t bin the data, as you don’t want to split up the data even more.</p>
<p>Now we run gradientForest with 10 bootstrapped trees to be generated by randomForest. This will take about 5 minutes.</p>
<div class="sourceCode" id="cb315"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb315-1"><a href="geassoc.html#cb315-1" aria-hidden="true" tabindex="-1"></a>wiflforest<span class="ot">=</span><span class="fu">gradientForest</span>(<span class="fu">cbind</span>(Ewifl,Gwifl), <span class="at">predictor.vars=</span>preds, <span class="at">response.vars=</span>specs, <span class="at">ntree=</span><span class="dv">10</span>, <span class="at">transform =</span> <span class="cn">NULL</span>, <span class="at">compact=</span>T, <span class="at">nbin=</span><span class="dv">101</span>, <span class="at">maxLevel=</span>lev, <span class="at">trace=</span>T)</span></code></pre></div>
<pre><code>## Calculating forests for 10000 species
## ......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## ....</code></pre>
<pre><code>## Warning in randomForest.default(x = X, y = spec_vec,
## maxLevel = maxLevel, : The response has five or fewer
## unique values. Are you sure you want to do regression?</code></pre>
<pre><code>## ...................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## ..................................</code></pre>
<pre><code>## Warning in randomForest.default(x = X, y = spec_vec,
## maxLevel = maxLevel, : The response has five or fewer
## unique values. Are you sure you want to do regression?</code></pre>
<pre><code>## .....................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## ..............</code></pre>
<pre><code>## Warning in randomForest.default(x = X, y = spec_vec,
## maxLevel = maxLevel, : The response has five or fewer
## unique values. Are you sure you want to do regression?</code></pre>
<pre><code>## .........................................
## ...........................</code></pre>
<pre><code>## Warning in randomForest.default(x = X, y = spec_vec,
## maxLevel = maxLevel, : The response has five or fewer
## unique values. Are you sure you want to do regression?</code></pre>
<pre><code>## ............................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## .......................................................
## ..............................................</code></pre>
<p>Now we can ask which environmental variables are significantly associated with the allelic structure of Willow Flycatchers (or at least the subset we are looking at).</p>
<p>This first plot and most useful from gradientForest is the predictor overall importance plot, which shows a simple barplot of the ranked importances of the physical variables. This shows the mean accuracy importance and the mean importance weighted by SNP R^2. The most reliable importances are the R^2 weighted importances.</p>
<div class="sourceCode" id="cb325"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb325-1"><a href="geassoc.html#cb325-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(wiflforest,<span class="at">plot.type=</span><span class="st">&quot;Overall.Importance&quot;</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:unnamed-chunk-192"></span>
<img src="merida-workshop-2022_files/figure-html/unnamed-chunk-192-1.png" alt="Overall importance of predictor variables" width="672" />
<p class="caption">
FIGURE 7.2: Overall importance of predictor variables
</p>
</div>
<p><strong>For the Willow Flycatcher subsetted data in this tutorial, what are the top 5 variables?</strong></p>
<p>Use the code below to look at Cuumulative importance, aka which variants are strongly associated with the top two variables. The cumulative importance is plotted separately for all species and averaged over all species. The cumulative importance can be interpreted as a mapping from an environmental gradient on to biological gradient.</p>
<div class="sourceCode" id="cb326"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb326-1"><a href="geassoc.html#cb326-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(wiflforest,<span class="at">plot.type=</span><span class="st">&quot;Cumulative.Importance&quot;</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:unnamed-chunk-193"></span>
<img src="merida-workshop-2022_files/figure-html/unnamed-chunk-193-1.png" alt="Individual and cumulative importance of SNPs associated with top 2 environmental variables" width="672" />
<p class="caption">
FIGURE 7.3: Individual and cumulative importance of SNPs associated with top 2 environmental variables
</p>
</div>
<p><strong>Which SNP has the greatest importance?</strong> For plotting purposes, only the first two top environmental variables are plotted. So the plots on the left are variants and turnover functions associated with Bio4, and the plots on the right are associated Bio3.</p>
<p>You can confirm this by looking at the range of your top environemntal variables and comparing it to the x-axis in the plots above using this code below:</p>
<div class="sourceCode" id="cb327"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb327-1"><a href="geassoc.html#cb327-1" aria-hidden="true" tabindex="-1"></a>Ewifl <span class="sc">%&gt;%</span> dplyr<span class="sc">::</span><span class="fu">select</span>(BIO4,BIO3) <span class="sc">%&gt;%</span> <span class="fu">summarise</span>(<span class="at">rangeBio4=</span><span class="fu">range</span>(BIO4),<span class="at">rangeBio3=</span><span class="fu">range</span>(BIO3))</span></code></pre></div>
<pre><code>## # A tibble: 2 × 2
##   rangeBio4 rangeBio3
##       &lt;dbl&gt;     &lt;dbl&gt;
## 1      3298        27
## 2     11946        54</code></pre>
<div class="sourceCode" id="cb329"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb329-1"><a href="geassoc.html#cb329-1" aria-hidden="true" tabindex="-1"></a><span class="co">#or </span></span>
<span id="cb329-2"><a href="geassoc.html#cb329-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb329-3"><a href="geassoc.html#cb329-3" aria-hidden="true" tabindex="-1"></a>Ewifl <span class="sc">%&gt;%</span> <span class="fu">pull</span>(BIO4) <span class="sc">%&gt;%</span> <span class="fu">range</span>()</span></code></pre></div>
<pre><code>## [1]  3298 11946</code></pre>
<div class="sourceCode" id="cb331"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb331-1"><a href="geassoc.html#cb331-1" aria-hidden="true" tabindex="-1"></a>Ewifl <span class="sc">%&gt;%</span> <span class="fu">pull</span>(BIO3) <span class="sc">%&gt;%</span> <span class="fu">range</span>()</span></code></pre></div>
<pre><code>## [1] 27 54</code></pre>
<p>Note, while gradient forest can look at variants associated with environmental variables like this, we tend to choose the top uncorrelated environmental variables and run latent factor mixed models to identify candidate genes associated with each environmental variable. Why? In gradient forest, we “bin” the variation, so we are looking for large patterns, not individual variants.</p>
</div>
<div id="basic-gradient-forest-preparation" class="section level2" number="7.5">
<h2><span class="header-section-number">7.5</span> Basic Gradient Forest Preparation</h2>
<p>So now that you see how gradient forest is run, let’s plot the results in space.</p>
<p>Gradientforest here was based off of 21 locations in the WIFL range. I’ve randomly chosen 100,000 points within the WIFL range and gotten the evironmental data for these 100,000 sites, which will allow us to have a more continuous view of important environmental associations in WIFL space. Load your 100,000 random points with associated environmental predictors. This can be created with another script that was included with the tutorial material (bioclim&amp;randoms.R), but we’ll use the ready-made grid created specifically for Willow Flycatchers.</p>
<div class="sourceCode" id="cb333"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb333-1"><a href="geassoc.html#cb333-1" aria-hidden="true" tabindex="-1"></a><span class="co">#WIFL current random points, filter out &#39;bad&#39; points</span></span>
<span id="cb333-2"><a href="geassoc.html#cb333-2" aria-hidden="true" tabindex="-1"></a>birdgrid<span class="ot">&lt;-</span> <span class="fu">read_csv</span>(<span class="st">&quot;data/6.0/wiflcurrentRH.csv&quot;</span>) <span class="sc">%&gt;%</span> <span class="co">#load your  random points with assocaited env predictors</span></span>
<span id="cb333-3"><a href="geassoc.html#cb333-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(BIO4<span class="sc">!=</span><span class="st">&quot;-9999&quot;</span>) <span class="sc">%&gt;%</span>  <span class="co">#remove the &quot;bad data&quot;</span></span>
<span id="cb333-4"><a href="geassoc.html#cb333-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>() <span class="co">#create the data frame</span></span></code></pre></div>
<pre><code>## Rows: 100000 Columns: 12</code></pre>
<pre><code>## ── Column specification ───────────────────────────────
## Delimiter: &quot;,&quot;
## dbl (12): X, Y, BIO2, BIO3, BIO4, BIO5, BIO8, BIO9,...</code></pre>
<pre><code>## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div class="sourceCode" id="cb337"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb337-1"><a href="geassoc.html#cb337-1" aria-hidden="true" tabindex="-1"></a><span class="do">##Let&#39;s look at bird grid, untransformed environmental variable</span></span>
<span id="cb337-2"><a href="geassoc.html#cb337-2" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(birdgrid)</span></code></pre></div>
<pre><code>##         X     Y BIO2 BIO3 BIO4 BIO5 BIO8 BIO9 BIO11
## 1  -76.17 44.09  103   26 9755  260   93   -5   -60
## 2 -103.38 42.94  164   36 9842  320  183  -43   -43
## 3  -69.11 44.32  113   29 9282  261   28  187   -54
## 4 -109.13 33.49  197   52 6719  287  175  121     9
## 5 -106.01 33.40  169   46 7305  314  214   84    37
## 6 -119.89 48.97  102   32 7206  167  -92   52   -98
##   BIO15 BIO17 BIO18
## 1    14   201   224
## 2    64    32   167
## 3    14   256   256
## 4    60    47   200
## 5    65    42   156
## 6    25   105   117</code></pre>
<div class="sourceCode" id="cb339"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb339-1"><a href="geassoc.html#cb339-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(birdgrid)</span></code></pre></div>
<pre><code>## [1] 99541    12</code></pre>
<p>We don’t have to use all the predictors in the next steps, in this case, we’ll use the top 4 uncorrelated variables. These variables were selected by moving down the list of ranked importance for the full model and discarding variables highly correlated (|Pearson’s r| &gt; 0.7) with a variable of higher importance. You can see the correlations among environmental variables across your sample sites. While it is clearly hard to read here, if you export this and open in excel, you can color code and keep track of the correlations among the ranked variables.</p>
<div class="sourceCode" id="cb341"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb341-1"><a href="geassoc.html#cb341-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(<span class="fu">cor</span>(Ewifl,<span class="at">method=</span><span class="st">&quot;pearson&quot;</span>))</span></code></pre></div>
<pre><code>##          BIO1     BIO2     BIO3    BIO4     BIO5
## BIO1  1.00000  0.06378  0.41991 -0.3227  0.70423
## BIO2  0.06378  1.00000  0.20607  0.2063  0.51230
## BIO3  0.41991  0.20607  1.00000 -0.8978 -0.07301
## BIO4 -0.32265  0.20626 -0.89779  1.0000  0.33424
## BIO5  0.70423  0.51230 -0.07301  0.3342  1.00000
## BIO6  0.81738 -0.18653  0.73399 -0.7881  0.22968
##         BIO6    BIO7     BIO8    BIO9    BIO10   BIO11
## BIO1  0.8174 -0.3214  0.41036  0.6098  0.83681  0.8838
## BIO2 -0.1865  0.4967 -0.07870  0.2901  0.21572 -0.0240
## BIO3  0.7340 -0.7323 -0.33935  0.7733 -0.07723  0.7565
## BIO4 -0.7881  0.9472  0.39863 -0.6709  0.24630 -0.7257
## BIO5  0.2297  0.4142  0.46587  0.3243  0.92631  0.3598
## BIO6  1.0000 -0.7907  0.02865  0.7580  0.37954  0.9787
##         BIO12     BIO13   BIO14    BIO15    BIO16
## BIO1 -0.06788 -0.009306 -0.3315  0.47177 -0.01376
## BIO2 -0.62811 -0.527137 -0.5750  0.03587 -0.54400
## BIO3  0.11332  0.283245 -0.4859  0.62916  0.27646
## BIO4 -0.43212 -0.542562  0.1622 -0.52172 -0.54341
## BIO5 -0.49614 -0.468213 -0.4053  0.10623 -0.47815
## BIO6  0.29580  0.381466 -0.2247  0.56762  0.38053
##        BIO17   BIO18    BIO19 NDVI_Mean NDVI_StDev
## BIO1 -0.3092 -0.3012  0.05711  -0.25183    -0.3159
## BIO2 -0.6153 -0.3973 -0.50485  -0.71683    -0.5122
## BIO3 -0.4319 -0.6322  0.38306  -0.12471    -0.3679
## BIO4  0.0850  0.4621 -0.64819  -0.19234     0.1161
## BIO5 -0.4351 -0.2278 -0.43996  -0.57552    -0.3845
## BIO6 -0.1567 -0.4205  0.48703   0.03524    -0.1907
##          SRTM TreeCover
## BIO1 -0.55867  -0.22919
## BIO2  0.62444  -0.31430
## BIO3  0.04009   0.04665
## BIO4  0.17262  -0.19766
## BIO5 -0.19354  -0.41770
## BIO6 -0.54098   0.01015</code></pre>
<p>For the remaining time in this tutorial we are focusing on the uncorrelated variables determined for the <strong>full</strong> Willow Flycatcher dataset (Ruegg et al. 2017), BIO11 (Mean temperature of coldest quarter), BIO5 (Max temperature of warmest month), BIO4 (Temperature seasonality (standard deviation *100)), and BIO17 (Precipitation of driest quarter).</p>
<p>Predict across your random sites using your Forest model:</p>
<div class="sourceCode" id="cb343"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb343-1"><a href="geassoc.html#cb343-1" aria-hidden="true" tabindex="-1"></a>predictors_uncor<span class="ot">&lt;-</span><span class="fu">c</span>(<span class="st">&quot;BIO11&quot;</span>,<span class="st">&quot;BIO5&quot;</span>,<span class="st">&quot;BIO4&quot;</span> ,<span class="st">&quot;BIO17&quot;</span>)</span>
<span id="cb343-2"><a href="geassoc.html#cb343-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb343-3"><a href="geassoc.html#cb343-3" aria-hidden="true" tabindex="-1"></a>currentgrid <span class="ot">&lt;-</span><span class="fu">cbind</span>(birdgrid[,<span class="fu">c</span>(<span class="st">&quot;X&quot;</span>,<span class="st">&quot;Y&quot;</span>)], <span class="fu">predict</span>(wiflforest,birdgrid[,predictors_uncor])) <span class="co">#predict across your random sites using your forest</span></span>
<span id="cb343-4"><a href="geassoc.html#cb343-4" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb343-5"><a href="geassoc.html#cb343-5" aria-hidden="true" tabindex="-1"></a><span class="do">##Let&#39;s look at currentgrid, transformed environmental variables</span></span>
<span id="cb343-6"><a href="geassoc.html#cb343-6" aria-hidden="true" tabindex="-1"></a><span class="do">##Notice how they were transformed</span></span>
<span id="cb343-7"><a href="geassoc.html#cb343-7" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(currentgrid)</span></code></pre></div>
<pre><code>##         X     Y      BIO11      BIO5     BIO4
## 1  -76.17 44.09  0.0004528  0.001014 0.022466
## 2 -103.38 42.94  0.0013002  0.002946 0.022643
## 3  -69.11 44.32  0.0005108  0.001068 0.021303
## 4 -109.13 33.49  0.0050652  0.002272 0.003932
## 5 -106.01 33.40  0.0064085  0.002934 0.004638
## 6 -119.89 48.97 -0.0001678 -0.002423 0.004499
##       BIO17
## 1 0.0049102
## 2 0.0007953
## 3 0.0057301
## 4 0.0014202
## 5 0.0013088
## 6 0.0025174</code></pre>
<div class="sourceCode" id="cb345"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb345-1"><a href="geassoc.html#cb345-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dim</span>(currentgrid)</span></code></pre></div>
<pre><code>## [1] 99541     6</code></pre>
<p>We can summarize the variation using principal components. We then convert those PC axes to RGB values for plotting.</p>
<div class="sourceCode" id="cb347"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb347-1"><a href="geassoc.html#cb347-1" aria-hidden="true" tabindex="-1"></a><span class="co">#Getting the PCs and Converting Them to RGB Color Values</span></span>
<span id="cb347-2"><a href="geassoc.html#cb347-2" aria-hidden="true" tabindex="-1"></a>PCs<span class="ot">=</span><span class="fu">prcomp</span>(currentgrid[,<span class="dv">3</span><span class="sc">:</span><span class="dv">6</span>]) <span class="co">#run your PCs on your top four variables</span></span>
<span id="cb347-3"><a href="geassoc.html#cb347-3" aria-hidden="true" tabindex="-1"></a><span class="co"># set up a colour palette for the mapping</span></span>
<span id="cb347-4"><a href="geassoc.html#cb347-4" aria-hidden="true" tabindex="-1"></a>a1 <span class="ot">&lt;-</span> PCs<span class="sc">$</span>x[,<span class="dv">1</span>]</span>
<span id="cb347-5"><a href="geassoc.html#cb347-5" aria-hidden="true" tabindex="-1"></a>a2 <span class="ot">&lt;-</span> PCs<span class="sc">$</span>x[,<span class="dv">2</span>]</span>
<span id="cb347-6"><a href="geassoc.html#cb347-6" aria-hidden="true" tabindex="-1"></a>a3 <span class="ot">&lt;-</span> PCs<span class="sc">$</span>x[,<span class="dv">3</span>]</span>
<span id="cb347-7"><a href="geassoc.html#cb347-7" aria-hidden="true" tabindex="-1"></a>r <span class="ot">&lt;-</span> a1<span class="sc">+</span>a2</span>
<span id="cb347-8"><a href="geassoc.html#cb347-8" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> <span class="sc">-</span>a2</span>
<span id="cb347-9"><a href="geassoc.html#cb347-9" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> a3<span class="sc">+</span>a2<span class="sc">-</span>a1</span>
<span id="cb347-10"><a href="geassoc.html#cb347-10" aria-hidden="true" tabindex="-1"></a>r <span class="ot">&lt;-</span> (r<span class="sc">-</span><span class="fu">min</span>(r)) <span class="sc">/</span> (<span class="fu">max</span>(r)<span class="sc">-</span><span class="fu">min</span>(r)) <span class="sc">*</span> <span class="dv">255</span></span>
<span id="cb347-11"><a href="geassoc.html#cb347-11" aria-hidden="true" tabindex="-1"></a>g <span class="ot">&lt;-</span> (g<span class="sc">-</span><span class="fu">min</span>(g)) <span class="sc">/</span> (<span class="fu">max</span>(g)<span class="sc">-</span><span class="fu">min</span>(g)) <span class="sc">*</span> <span class="dv">255</span></span>
<span id="cb347-12"><a href="geassoc.html#cb347-12" aria-hidden="true" tabindex="-1"></a>b <span class="ot">&lt;-</span> (b<span class="sc">-</span><span class="fu">min</span>(b)) <span class="sc">/</span> (<span class="fu">max</span>(b)<span class="sc">-</span><span class="fu">min</span>(b)) <span class="sc">*</span> <span class="dv">255</span></span>
<span id="cb347-13"><a href="geassoc.html#cb347-13" aria-hidden="true" tabindex="-1"></a>wiflcols<span class="ot">=</span><span class="fu">rgb</span>(r,g,b,<span class="at">max=</span><span class="dv">255</span>)</span>
<span id="cb347-14"><a href="geassoc.html#cb347-14" aria-hidden="true" tabindex="-1"></a>wiflcols2<span class="ot">=</span><span class="fu">col2rgb</span>(wiflcols)</span>
<span id="cb347-15"><a href="geassoc.html#cb347-15" aria-hidden="true" tabindex="-1"></a>wiflcols3<span class="ot">=</span><span class="fu">t</span>(wiflcols2)</span>
<span id="cb347-16"><a href="geassoc.html#cb347-16" aria-hidden="true" tabindex="-1"></a>gradients<span class="ot">=</span><span class="fu">cbind</span>(currentgrid[<span class="fu">c</span>(<span class="st">&quot;X&quot;</span>,<span class="st">&quot;Y&quot;</span>)],wiflcols3)</span></code></pre></div>
<p>The multi-dimensional biological space can be plotted by taking the principle component of the transformed grid and presenting the first two dimensions in a biplot. You’ve already defined the PCs above and the RGB color palette. Different coordinate positions in the biplot represent differing allele frequency compositions as associated with the predictors.</p>
<div class="sourceCode" id="cb348"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb348-1"><a href="geassoc.html#cb348-1" aria-hidden="true" tabindex="-1"></a><span class="do">##Biplot specifics</span></span>
<span id="cb348-2"><a href="geassoc.html#cb348-2" aria-hidden="true" tabindex="-1"></a>nvs <span class="ot">&lt;-</span> <span class="fu">dim</span>(PCs<span class="sc">$</span>rotation)[<span class="dv">1</span>]</span>
<span id="cb348-3"><a href="geassoc.html#cb348-3" aria-hidden="true" tabindex="-1"></a>vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;BIO11&quot;</span>,<span class="st">&quot;BIO5&quot;</span>,<span class="st">&quot;BIO4&quot;</span> ,<span class="st">&quot;BIO17&quot;</span>)</span>
<span id="cb348-4"><a href="geassoc.html#cb348-4" aria-hidden="true" tabindex="-1"></a>lv <span class="ot">&lt;-</span> <span class="fu">length</span>(vec)</span>
<span id="cb348-5"><a href="geassoc.html#cb348-5" aria-hidden="true" tabindex="-1"></a>vind <span class="ot">&lt;-</span> <span class="fu">rownames</span>(PCs<span class="sc">$</span>rotation) <span class="sc">%in%</span> vec</span>
<span id="cb348-6"><a href="geassoc.html#cb348-6" aria-hidden="true" tabindex="-1"></a>scal <span class="ot">&lt;-</span><span class="dv">100</span></span>
<span id="cb348-7"><a href="geassoc.html#cb348-7" aria-hidden="true" tabindex="-1"></a>xrng <span class="ot">&lt;-</span> <span class="fu">range</span>(PCs<span class="sc">$</span>x[, <span class="dv">1</span>], PCs<span class="sc">$</span>rotation[, <span class="dv">1</span>]<span class="sc">/</span>scal) <span class="sc">*</span><span class="fl">1.2</span></span>
<span id="cb348-8"><a href="geassoc.html#cb348-8" aria-hidden="true" tabindex="-1"></a>yrng <span class="ot">&lt;-</span> <span class="fu">range</span>(PCs<span class="sc">$</span>x[, <span class="dv">2</span>], PCs<span class="sc">$</span>rotation[, <span class="dv">2</span>]<span class="sc">/</span>scal) <span class="sc">*</span><span class="fl">1.2</span></span>
<span id="cb348-9"><a href="geassoc.html#cb348-9" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>((PCs<span class="sc">$</span>x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]), <span class="at">xlim =</span> xrng, <span class="at">ylim =</span> yrng, <span class="at">pch =</span> <span class="st">&quot;.&quot;</span>, <span class="at">cex =</span> <span class="dv">4</span>, <span class="at">col =</span> <span class="fu">rgb</span>(r, g, b, <span class="at">max =</span> <span class="dv">255</span>), <span class="at">asp =</span> <span class="dv">1</span>)</span>
<span id="cb348-10"><a href="geassoc.html#cb348-10" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(PCs<span class="sc">$</span>rotation[<span class="sc">!</span>vind, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]<span class="sc">/</span>scal, <span class="at">pch =</span> <span class="st">&quot;+&quot;</span>)</span>
<span id="cb348-11"><a href="geassoc.html#cb348-11" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(<span class="fu">rep</span>(<span class="dv">0</span>, lv), <span class="fu">rep</span>(<span class="dv">0</span>, lv), PCs<span class="sc">$</span>rotation[vec,<span class="dv">1</span>]<span class="sc">/</span>scal, PCs<span class="sc">$</span>rotation[vec, <span class="dv">2</span>]<span class="sc">/</span>scal, <span class="at">length =</span> <span class="fl">0.0625</span>)</span>
<span id="cb348-12"><a href="geassoc.html#cb348-12" aria-hidden="true" tabindex="-1"></a>jit <span class="ot">&lt;-</span> <span class="fl">0.0015</span></span>
<span id="cb348-13"><a href="geassoc.html#cb348-13" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(PCs<span class="sc">$</span>rotation[vec, <span class="dv">1</span>]<span class="sc">/</span>scal <span class="sc">+</span> jit <span class="sc">*</span> <span class="fu">sign</span>(PCs<span class="sc">$</span>rotation[vec,<span class="dv">1</span>]), PCs<span class="sc">$</span>rotation[vec, <span class="dv">2</span>]<span class="sc">/</span>scal <span class="sc">+</span> jit <span class="sc">*</span> <span class="fu">sign</span>(PCs<span class="sc">$</span>rotation[vec,<span class="dv">2</span>]), <span class="at">labels =</span> vec)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:unnamed-chunk-199"></span>
<img src="merida-workshop-2022_files/figure-html/unnamed-chunk-199-1.png" alt="A principle component plot of the top 4 uncorrelated variables" width="672" />
<p class="caption">
FIGURE 7.4: A principle component plot of the top 4 uncorrelated variables
</p>
</div>
<p><strong>Save this image using the code chunk below. What environmental variables are associated with the colors green, pink and blue? Knowing this will help with the interpretation of plotting this in space.</strong></p>
<div class="sourceCode" id="cb349"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb349-1"><a href="geassoc.html#cb349-1" aria-hidden="true" tabindex="-1"></a><span class="fu">dir.create</span>(<span class="st">&quot;stored_results/6.0&quot;</span>, <span class="at">showWarnings =</span> <span class="cn">FALSE</span>, <span class="at">recursive =</span> <span class="cn">TRUE</span>)</span>
<span id="cb349-2"><a href="geassoc.html#cb349-2" aria-hidden="true" tabindex="-1"></a><span class="fu">pdf</span>(<span class="st">&quot;stored_results/6.0/WIFL.PCA_uncorrelated.pdf&quot;</span>)</span>
<span id="cb349-3"><a href="geassoc.html#cb349-3" aria-hidden="true" tabindex="-1"></a>nvs <span class="ot">&lt;-</span> <span class="fu">dim</span>(PCs<span class="sc">$</span>rotation)[<span class="dv">1</span>]</span>
<span id="cb349-4"><a href="geassoc.html#cb349-4" aria-hidden="true" tabindex="-1"></a>vec <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;BIO11&quot;</span>,<span class="st">&quot;BIO5&quot;</span>,<span class="st">&quot;BIO4&quot;</span> ,<span class="st">&quot;BIO17&quot;</span>)</span>
<span id="cb349-5"><a href="geassoc.html#cb349-5" aria-hidden="true" tabindex="-1"></a>lv <span class="ot">&lt;-</span> <span class="fu">length</span>(vec)</span>
<span id="cb349-6"><a href="geassoc.html#cb349-6" aria-hidden="true" tabindex="-1"></a>vind <span class="ot">&lt;-</span> <span class="fu">rownames</span>(PCs<span class="sc">$</span>rotation) <span class="sc">%in%</span> vec</span>
<span id="cb349-7"><a href="geassoc.html#cb349-7" aria-hidden="true" tabindex="-1"></a>scal <span class="ot">&lt;-</span><span class="dv">100</span></span>
<span id="cb349-8"><a href="geassoc.html#cb349-8" aria-hidden="true" tabindex="-1"></a>xrng <span class="ot">&lt;-</span> <span class="fu">range</span>(PCs<span class="sc">$</span>x[, <span class="dv">1</span>], PCs<span class="sc">$</span>rotation[, <span class="dv">1</span>]<span class="sc">/</span>scal) <span class="sc">*</span><span class="fl">1.2</span></span>
<span id="cb349-9"><a href="geassoc.html#cb349-9" aria-hidden="true" tabindex="-1"></a>yrng <span class="ot">&lt;-</span> <span class="fu">range</span>(PCs<span class="sc">$</span>x[, <span class="dv">2</span>], PCs<span class="sc">$</span>rotation[, <span class="dv">2</span>]<span class="sc">/</span>scal) <span class="sc">*</span><span class="fl">1.2</span></span>
<span id="cb349-10"><a href="geassoc.html#cb349-10" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>((PCs<span class="sc">$</span>x[, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]), <span class="at">xlim =</span> xrng, <span class="at">ylim =</span> yrng, <span class="at">pch =</span> <span class="st">&quot;.&quot;</span>, <span class="at">cex =</span> <span class="dv">4</span>, <span class="at">col =</span> <span class="fu">rgb</span>(r, g, b, <span class="at">max =</span> <span class="dv">255</span>), <span class="at">asp =</span> <span class="dv">1</span>)</span>
<span id="cb349-11"><a href="geassoc.html#cb349-11" aria-hidden="true" tabindex="-1"></a><span class="fu">points</span>(PCs<span class="sc">$</span>rotation[<span class="sc">!</span>vind, <span class="dv">1</span><span class="sc">:</span><span class="dv">2</span>]<span class="sc">/</span>scal, <span class="at">pch =</span> <span class="st">&quot;+&quot;</span>)</span>
<span id="cb349-12"><a href="geassoc.html#cb349-12" aria-hidden="true" tabindex="-1"></a><span class="fu">arrows</span>(<span class="fu">rep</span>(<span class="dv">0</span>, lv), <span class="fu">rep</span>(<span class="dv">0</span>, lv), PCs<span class="sc">$</span>rotation[vec,<span class="dv">1</span>]<span class="sc">/</span>scal, PCs<span class="sc">$</span>rotation[vec, <span class="dv">2</span>]<span class="sc">/</span>scal, <span class="at">length =</span> <span class="fl">0.0625</span>)</span>
<span id="cb349-13"><a href="geassoc.html#cb349-13" aria-hidden="true" tabindex="-1"></a>jit <span class="ot">&lt;-</span> <span class="fl">0.0015</span></span>
<span id="cb349-14"><a href="geassoc.html#cb349-14" aria-hidden="true" tabindex="-1"></a><span class="fu">text</span>(PCs<span class="sc">$</span>rotation[vec, <span class="dv">1</span>]<span class="sc">/</span>scal <span class="sc">+</span> jit <span class="sc">*</span> <span class="fu">sign</span>(PCs<span class="sc">$</span>rotation[vec,<span class="dv">1</span>]), PCs<span class="sc">$</span>rotation[vec, <span class="dv">2</span>]<span class="sc">/</span>scal <span class="sc">+</span> jit <span class="sc">*</span> <span class="fu">sign</span>(PCs<span class="sc">$</span>rotation[vec,<span class="dv">2</span>]), <span class="at">labels =</span> vec)</span>
<span id="cb349-15"><a href="geassoc.html#cb349-15" aria-hidden="true" tabindex="-1"></a><span class="fu">dev.off</span>()</span></code></pre></div>
<pre><code>## quartz_off_screen 
##                 2</code></pre>
<p>With this, we can draw our first gradient forest map:</p>
<div class="sourceCode" id="cb351"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb351-1"><a href="geassoc.html#cb351-1" aria-hidden="true" tabindex="-1"></a>wiflcols <span class="ot">&lt;-</span> <span class="fu">rgb</span>(r,g,b,<span class="at">max=</span><span class="dv">255</span>)</span>
<span id="cb351-2"><a href="geassoc.html#cb351-2" aria-hidden="true" tabindex="-1"></a>wiflcols2 <span class="ot">&lt;-</span> <span class="fu">col2rgb</span>(wiflcols)</span>
<span id="cb351-3"><a href="geassoc.html#cb351-3" aria-hidden="true" tabindex="-1"></a>wiflcols3<span class="ot">&lt;-</span> <span class="fu">t</span>(wiflcols2)</span>
<span id="cb351-4"><a href="geassoc.html#cb351-4" aria-hidden="true" tabindex="-1"></a>gradients<span class="ot">&lt;-</span><span class="fu">cbind</span>(currentgrid[<span class="fu">c</span>(<span class="st">&quot;X&quot;</span>,<span class="st">&quot;Y&quot;</span>)],wiflcols3)</span>
<span id="cb351-5"><a href="geassoc.html#cb351-5" aria-hidden="true" tabindex="-1"></a>wiflmap<span class="ot">&lt;-</span>gradients</span>
<span id="cb351-6"><a href="geassoc.html#cb351-6" aria-hidden="true" tabindex="-1"></a><span class="fu">coordinates</span>(wiflmap)<span class="ot">&lt;-</span> <span class="er">~</span>X<span class="sc">+</span>Y <span class="co">#setting the coordinates</span></span>
<span id="cb351-7"><a href="geassoc.html#cb351-7" aria-hidden="true" tabindex="-1"></a><span class="fu">proj4string</span>(wiflmap) <span class="ot">&lt;-</span> <span class="fu">CRS</span>(<span class="st">&quot;+proj=longlat +datum=WGS84&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb352"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb352-1"><a href="geassoc.html#cb352-1" aria-hidden="true" tabindex="-1"></a><span class="fu">par</span>(<span class="at">mar=</span><span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>,<span class="dv">0</span>))</span>
<span id="cb352-2"><a href="geassoc.html#cb352-2" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(wiflmap,<span class="at">col=</span><span class="fu">rgb</span>(r,g,b,<span class="at">max=</span><span class="dv">255</span>),<span class="at">pch=</span><span class="dv">15</span>)</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:unnamed-chunk-202"></span>
<img src="merida-workshop-2022_files/figure-html/unnamed-chunk-202-1.png" alt="Our gradient forest model of SNP variation in the Willow Flycatcher breeding range" width="672" />
<p class="caption">
FIGURE 7.5: Our gradient forest model of SNP variation in the Willow Flycatcher breeding range
</p>
</div>
<p><strong>How would you interpret the different colors on thie WIFL map? Is the endangered southwest Willow Flycatcher locally adapted? </strong></p>
<p>Let’s plot this another way, on an actual map of North America. To do this, we need some additional shapefiles.</p>
<p>The domain here allows you to crop the shapefiles to just your species range.</p>
<div class="sourceCode" id="cb353"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb353-1"><a href="geassoc.html#cb353-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(sf)</span>
<span id="cb353-2"><a href="geassoc.html#cb353-2" aria-hidden="true" tabindex="-1"></a>coastlines <span class="ot">&lt;-</span> <span class="fu">st_read</span>(<span class="st">&quot;data/6.0/ne_shapefiles/ne_10m_coastline.shp&quot;</span>)</span></code></pre></div>
<pre><code>## Reading layer `ne_10m_coastline&#39; from data source 
##   `/Users/eriq/Documents/git-repos/merida-workshop-2022/data/6.0/ne_shapefiles/ne_10m_coastline.shp&#39; 
##   using driver `ESRI Shapefile&#39;
## Simple feature collection with 4133 features and 3 fields
## Geometry type: LINESTRING
## Dimension:     XY
## Bounding box:  xmin: -180 ymin: -85.22 xmax: 180 ymax: 83.63
## Geodetic CRS:  WGS 84</code></pre>
<div class="sourceCode" id="cb355"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb355-1"><a href="geassoc.html#cb355-1" aria-hidden="true" tabindex="-1"></a>domain <span class="ot">&lt;-</span> <span class="fu">c</span>(</span>
<span id="cb355-2"><a href="geassoc.html#cb355-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">xmin =</span> <span class="sc">-</span><span class="dv">140</span>, </span>
<span id="cb355-3"><a href="geassoc.html#cb355-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">xmax =</span> <span class="sc">-</span><span class="dv">55</span>,</span>
<span id="cb355-4"><a href="geassoc.html#cb355-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">ymin =</span> <span class="dv">15</span>,</span>
<span id="cb355-5"><a href="geassoc.html#cb355-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">ymax =</span> <span class="dv">50</span></span>
<span id="cb355-6"><a href="geassoc.html#cb355-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb355-7"><a href="geassoc.html#cb355-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb355-8"><a href="geassoc.html#cb355-8" aria-hidden="true" tabindex="-1"></a>coast_cropped <span class="ot">&lt;-</span> <span class="fu">st_crop</span>(coastlines, domain)</span></code></pre></div>
<pre><code>## Warning: attribute variables are assumed to be
## spatially constant throughout all geometries</code></pre>
<p>This code allows you to plot, but also change the domain to what will encompass your species shapefile. Suffice it to say I went through a fair number of iterations to ge it right.</p>
<p>Once you have the shapefiles and cropped domain, you can create a map, project the map, and then crop it to what region you want. Using the above code, the domain can/should be larger than what you will be plotting. You can further crop later!</p>
<div class="sourceCode" id="cb357"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb357-1"><a href="geassoc.html#cb357-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(ggspatial)</span>
<span id="cb357-2"><a href="geassoc.html#cb357-2" aria-hidden="true" tabindex="-1"></a>countries_cropped <span class="ot">&lt;-</span>  <span class="fu">st_read</span>(<span class="st">&quot;data/6.0/ne_shapefiles/ne_10m_admin_0_boundary_lines_land.shp&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb357-3"><a href="geassoc.html#cb357-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">st_crop</span>(domain)</span></code></pre></div>
<pre><code>## Reading layer `ne_10m_admin_0_boundary_lines_land&#39; from data source `/Users/eriq/Documents/git-repos/merida-workshop-2022/data/6.0/ne_shapefiles/ne_10m_admin_0_boundary_lines_land.shp&#39; 
##   using driver `ESRI Shapefile&#39;
## Simple feature collection with 462 features and 18 fields
## Geometry type: MULTILINESTRING
## Dimension:     XY
## Bounding box:  xmin: -141 ymin: -55.12 xmax: 141 ymax: 70.08
## Geodetic CRS:  WGS 84</code></pre>
<pre><code>## Warning: attribute variables are assumed to be
## spatially constant throughout all geometries</code></pre>
<div class="sourceCode" id="cb360"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb360-1"><a href="geassoc.html#cb360-1" aria-hidden="true" tabindex="-1"></a>states_cropped <span class="ot">&lt;-</span> <span class="fu">st_read</span>(<span class="st">&quot;data/6.0/ne_shapefiles/ne_10m_admin_1_states_provinces_lines.shp&quot;</span>) <span class="sc">%&gt;%</span></span>
<span id="cb360-2"><a href="geassoc.html#cb360-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">st_crop</span>(domain)</span></code></pre></div>
<pre><code>## Reading layer `ne_10m_admin_1_states_provinces_lines&#39; from data source `/Users/eriq/Documents/git-repos/merida-workshop-2022/data/6.0/ne_shapefiles/ne_10m_admin_1_states_provinces_lines.shp&#39; 
##   using driver `ESRI Shapefile&#39;
## Simple feature collection with 10114 features and 21 fields
## Geometry type: MULTILINESTRING
## Dimension:     XY
## Bounding box:  xmin: -178.1 ymin: -49.25 xmax: 178.4 ymax: 81.13
## Geodetic CRS:  WGS 84</code></pre>
<pre><code>## Warning: attribute variables are assumed to be
## spatially constant throughout all geometries</code></pre>
<div class="sourceCode" id="cb363"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb363-1"><a href="geassoc.html#cb363-1" aria-hidden="true" tabindex="-1"></a>mapg <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb363-2"><a href="geassoc.html#cb363-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> coast_cropped) <span class="sc">+</span></span>
<span id="cb363-3"><a href="geassoc.html#cb363-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> countries_cropped, <span class="at">fill =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb363-4"><a href="geassoc.html#cb363-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> states_cropped, <span class="at">fill =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb363-5"><a href="geassoc.html#cb363-5" aria-hidden="true" tabindex="-1"></a>  ggspatial<span class="sc">::</span><span class="fu">layer_spatial</span>(wiflmap,<span class="at">col=</span><span class="fu">rgb</span>(r,g,b,<span class="at">max=</span><span class="dv">255</span>),<span class="at">pch=</span><span class="dv">15</span>) <span class="sc">+</span></span>
<span id="cb363-6"><a href="geassoc.html#cb363-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data=</span>wifl_shape,<span class="at">fill=</span><span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb363-7"><a href="geassoc.html#cb363-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() </span>
<span id="cb363-8"><a href="geassoc.html#cb363-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb363-9"><a href="geassoc.html#cb363-9" aria-hidden="true" tabindex="-1"></a>mapg</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:unnamed-chunk-204"></span>
<img src="merida-workshop-2022_files/figure-html/unnamed-chunk-204-1.png" alt="Our gradient forest model of SNP variation in the Willow Flycatcher breeding range on a map" width="672" />
<p class="caption">
FIGURE 7.6: Our gradient forest model of SNP variation in the Willow Flycatcher breeding range on a map
</p>
</div>
<p>Now, let’s see how our developing map looks under such a projection. We just define <code>lamproj</code> as a coordinate reference system and pass it in to <code>coord_sf()</code></p>
<div class="sourceCode" id="cb364"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb364-1"><a href="geassoc.html#cb364-1" aria-hidden="true" tabindex="-1"></a>lamproj <span class="ot">&lt;-</span> <span class="st">&quot;+proj=lcc +lat_1=20 +lat_2=60 +lat_0=40 +lon_0=-100 +x_0=0 +y_0=0 +ellps=GRS80 +datum=NAD83 +units=m +no_defs&quot;</span></span>
<span id="cb364-2"><a href="geassoc.html#cb364-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb364-3"><a href="geassoc.html#cb364-3" aria-hidden="true" tabindex="-1"></a><span class="co">#Plotting with projection, and cutting out rectangle (Don&#39;t need a lot of extra space)</span></span>
<span id="cb364-4"><a href="geassoc.html#cb364-4" aria-hidden="true" tabindex="-1"></a>rectangled <span class="ot">&lt;-</span>mapg <span class="sc">+</span></span>
<span id="cb364-5"><a href="geassoc.html#cb364-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">xlab</span>(<span class="st">&quot;Longitude&quot;</span>) <span class="sc">+</span> <span class="fu">ylab</span>(<span class="st">&quot;Latitude&quot;</span>)<span class="sc">+</span></span>
<span id="cb364-6"><a href="geassoc.html#cb364-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">coord_sf</span>(</span>
<span id="cb364-7"><a href="geassoc.html#cb364-7" aria-hidden="true" tabindex="-1"></a>    <span class="at">crs =</span> <span class="fu">st_crs</span>(lamproj), </span>
<span id="cb364-8"><a href="geassoc.html#cb364-8" aria-hidden="true" tabindex="-1"></a>    <span class="at">xlim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="fl">2200476.8</span>, <span class="dv">2859510</span>), </span>
<span id="cb364-9"><a href="geassoc.html#cb364-9" aria-hidden="true" tabindex="-1"></a>    <span class="at">ylim =</span> <span class="fu">c</span>(<span class="sc">-</span><span class="dv">1899797</span>, <span class="fl">1809910.8</span>),</span>
<span id="cb364-10"><a href="geassoc.html#cb364-10" aria-hidden="true" tabindex="-1"></a>    <span class="at">expand =</span> <span class="cn">FALSE</span>) </span>
<span id="cb364-11"><a href="geassoc.html#cb364-11" aria-hidden="true" tabindex="-1"></a>rectangled</span></code></pre></div>
<p><img src="merida-workshop-2022_files/figure-html/unnamed-chunk-205-1.png" width="672" /></p>
<p>Rename this file. I put the plot I plotted in here, but let’s make sure this works for you!!</p>
<div class="sourceCode" id="cb365"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb365-1"><a href="geassoc.html#cb365-1" aria-hidden="true" tabindex="-1"></a><span class="fu">ggsave</span>(<span class="st">&quot;stored_results/6.0/WIFL.GF_map.pdf&quot;</span>)</span></code></pre></div>
<pre><code>## Saving 7 x 5 in image</code></pre>
</div>
<div id="defining-genomic-vulnerabilitygenomic-offset-given-future-climate-predictions" class="section level2" number="7.6">
<h2><span class="header-section-number">7.6</span> Defining “Genomic Vulnerability/Genomic offset” Given Future Climate Predictions</h2>
<p>The plots above are the genotype-environment relationship given present-day climate rasters. However, we know rapid fluctuations in temperature and precipitation associated with climate change can alter the suitability of particular regions making it necessary for individuals to either adapt, disperse or die if the changes are extreme enough. Those species that possess standing genetic variation for climate-related traits (i.e. have adaptive capacity) are most likely to have the ability to adapt to rapidly changing environments. Those species that unable to adapt are deemed most vulneraable.</p>
<p>To investigate which populations might be most vulnerable to future climate change, we defined the metric “genomic vulnerability” as the mismatch between current and predicted future genomic variation based on genotype-environment relationships modeled across contemporary populations. We followed the method presented in Fitzpatrick and Keller (2015) to calculate genomic vulnerability using an extension of the gradient forest analysis. Populations with the greatest mismatch are least likely to adapt quickly enough to track future climate shifts, potentially resulting in population declines or extirpations.</p>
<p>Here, we read in the future climate predictions for each random point. Since different amounts of heat-trapping gases released into the atmosphere by human activities produce different projected increases in Earth’s temperature, we provided two such predicted models: the 2.6 emission predictions and the 8.5 emission predictions for the year 2050.</p>
<p>The lowest recent emissions pathway (RCP), 2.6, assumes immediate and rapid reductions in emissions and would result in about 2.5°F of warming in this century. The highest emissions pathway, RCP 8.5, roughly similar to a continuation of the current path of global emissions increases, is projected to lead to more than 8°F warming by 2100, with a high-end possibility of more than 11°F. Each of these is a composite of multiple prediction models and have rather different predictions.</p>
<div class="sourceCode" id="cb367"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb367-1"><a href="geassoc.html#cb367-1" aria-hidden="true" tabindex="-1"></a>future1<span class="ot">&lt;-</span><span class="fu">read_csv</span>(<span class="st">&quot;data/6.0/wiflfuture2-6-2050.csv&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb367-2"><a href="geassoc.html#cb367-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(BIO4<span class="sc">!=</span><span class="st">&quot;-9999&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb367-3"><a href="geassoc.html#cb367-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span></code></pre></div>
<pre><code>## Rows: 100000 Columns: 12</code></pre>
<pre><code>## ── Column specification ───────────────────────────────
## Delimiter: &quot;,&quot;
## dbl (12): X, Y, BIO2, BIO3, BIO4, BIO5, BIO8, BIO9,...</code></pre>
<pre><code>## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div class="sourceCode" id="cb371"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb371-1"><a href="geassoc.html#cb371-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(future1)</span></code></pre></div>
<pre><code>##         X     Y BIO2 BIO3 BIO4 BIO5 BIO8 BIO9 BIO11
## 1  -76.17 44.09  102   26 9584  283   65  138   -34
## 2 -103.38 42.94  166   36 9942  342  171  -15   -24
## 3  -69.11 44.32  112   29 9209  282   42  173   -31
## 4 -109.13 33.49  202   51 6941  310  195  133    24
## 5 -106.01 33.40  174   46 7508  338  235  109    53
## 6 -119.89 48.97  101   31 7244  192  -70   82   -79
##   BIO15 BIO17 BIO18
## 1    16   209   221
## 2    63    35   155
## 3    17   259   273
## 4    69    33   202
## 5    71    30   153
## 6    33    97   107</code></pre>
<div class="sourceCode" id="cb373"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb373-1"><a href="geassoc.html#cb373-1" aria-hidden="true" tabindex="-1"></a>future2<span class="ot">&lt;-</span><span class="fu">read_csv</span>(<span class="st">&quot;data/6.0/wiflfuture8-5-2050.csv&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb373-2"><a href="geassoc.html#cb373-2" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(BIO4<span class="sc">!=</span><span class="st">&quot;-9999&quot;</span>) <span class="sc">%&gt;%</span> </span>
<span id="cb373-3"><a href="geassoc.html#cb373-3" aria-hidden="true" tabindex="-1"></a>  <span class="fu">as.data.frame</span>()</span></code></pre></div>
<pre><code>## Rows: 100000 Columns: 12</code></pre>
<pre><code>## ── Column specification ───────────────────────────────
## Delimiter: &quot;,&quot;
## dbl (12): X, Y, BIO2, BIO3, BIO4, BIO5, BIO8, BIO9,...</code></pre>
<pre><code>## 
## ℹ Use `spec()` to retrieve the full column specification for this data.
## ℹ Specify the column types or set `show_col_types = FALSE` to quiet this message.</code></pre>
<div class="sourceCode" id="cb377"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb377-1"><a href="geassoc.html#cb377-1" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(future2)</span></code></pre></div>
<pre><code>##         X     Y BIO2 BIO3  BIO4 BIO5 BIO8 BIO9 BIO11
## 1  -76.17 44.09  101   26  9652  299   46  175   -19
## 2 -103.38 42.94  166   35 10238  359  183   -6   -15
## 3  -69.11 44.32  111   29  9211  295   45  209   -18
## 4 -109.13 33.49  203   51  7078  323  208  142    33
## 5 -106.01 33.40  175   45  7602  351  249  127    63
## 6 -119.89 48.97  101   30  7521  210  -64  111   -70
##   BIO15 BIO17 BIO18
## 1    16   215   224
## 2    62    41   153
## 3    19   257   269
## 4    69    30   198
## 5    75    27   152
## 6    37    87    90</code></pre>
<p>We then use our gradient forest model to transform these into “biological space” just as we did for random points across contemporary environments.</p>
<div class="sourceCode" id="cb379"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb379-1"><a href="geassoc.html#cb379-1" aria-hidden="true" tabindex="-1"></a>futuregrid <span class="ot">&lt;-</span> <span class="fu">cbind</span>(future1[,<span class="fu">c</span>(<span class="st">&quot;X&quot;</span>,<span class="st">&quot;Y&quot;</span>)], <span class="fu">predict</span>(wiflforest,future1[,predictors_uncor]))</span></code></pre></div>
<p>Finally, we calculate genomic vulnerability. Mathematically, for a given point this is the <strong>Euclidean distance</strong> between the current biological space (currentgrid) and the future biological space (futuregrid):</p>
<div class="sourceCode" id="cb380"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb380-1"><a href="geassoc.html#cb380-1" aria-hidden="true" tabindex="-1"></a>coords<span class="ot">&lt;-</span>birdgrid[,<span class="fu">c</span>(<span class="st">&quot;X&quot;</span>,<span class="st">&quot;Y&quot;</span>)]</span>
<span id="cb380-2"><a href="geassoc.html#cb380-2" aria-hidden="true" tabindex="-1"></a>euc <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span><span class="cn">NA</span>,<span class="at">nrow=</span><span class="fu">nrow</span>(futuregrid),<span class="at">ncol=</span><span class="dv">3</span>)</span>
<span id="cb380-3"><a href="geassoc.html#cb380-3" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(currentgrid)) {</span>
<span id="cb380-4"><a href="geassoc.html#cb380-4" aria-hidden="true" tabindex="-1"></a>  euc[j,] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">as.numeric</span>(coords[j,]), </span>
<span id="cb380-5"><a href="geassoc.html#cb380-5" aria-hidden="true" tabindex="-1"></a>               <span class="fu">as.numeric</span>(<span class="fu">dist</span>(<span class="fu">rbind</span>(currentgrid[j,], futuregrid[j,]))))</span>
<span id="cb380-6"><a href="geassoc.html#cb380-6" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb380-7"><a href="geassoc.html#cb380-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb380-8"><a href="geassoc.html#cb380-8" aria-hidden="true" tabindex="-1"></a>euc <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(euc)</span>
<span id="cb380-9"><a href="geassoc.html#cb380-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(euc) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Long&quot;</span>,<span class="st">&quot;Lat&quot;</span>,<span class="st">&quot;Vulnerability&quot;</span>)</span>
<span id="cb380-10"><a href="geassoc.html#cb380-10" aria-hidden="true" tabindex="-1"></a><span class="fu">head</span>(euc)</span></code></pre></div>
<pre><code>##      Long   Lat Vulnerability
## 1  -76.17 44.09      0.001783
## 2 -103.38 42.94      0.001586
## 3  -69.11 44.32      0.001739
## 4 -109.13 33.49      0.001362
## 5 -106.01 33.40      0.002192
## 6 -119.89 48.97      0.001132</code></pre>
<p>This again takes a moment. Once we have this metric, we can easily map genomic vulnerability across the range. First by choosing a color palette, then by converting a dataframe of distances to a spatial polygon and giving it the same extent as the map it’s going to be plotted on and finally, mapping in ggplot.</p>
<div class="sourceCode" id="cb382"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb382-1"><a href="geassoc.html#cb382-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(scales)</span>
<span id="cb382-2"><a href="geassoc.html#cb382-2" aria-hidden="true" tabindex="-1"></a>colramp2<span class="ot">=</span><span class="fu">rev</span>(<span class="fu">brewer.pal</span>(<span class="at">n =</span> <span class="dv">9</span>, <span class="at">name =</span> <span class="st">&quot;RdYlBu&quot;</span>)[<span class="fu">c</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">9</span>)])</span>
<span id="cb382-3"><a href="geassoc.html#cb382-3" aria-hidden="true" tabindex="-1"></a>Futcolors<span class="ot">&lt;-</span><span class="fu">colorRampPalette</span>(colramp2)</span>
<span id="cb382-4"><a href="geassoc.html#cb382-4" aria-hidden="true" tabindex="-1"></a><span class="fu">show_col</span>(<span class="fu">Futcolors</span>(<span class="dv">100</span>))</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:unnamed-chunk-210"></span>
<img src="merida-workshop-2022_files/figure-html/unnamed-chunk-210-1.png" alt="Color Ramp for plotting" width="672" />
<p class="caption">
FIGURE 7.7: Color Ramp for plotting
</p>
</div>
<p>Notice that in the <em>rev</em> in the above code chunk will reverse the RColorBrewer color palette so the highest vulnerablity corresponds to the red color, rather than blue.</p>
<div class="sourceCode" id="cb383"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb383-1"><a href="geassoc.html#cb383-1" aria-hidden="true" tabindex="-1"></a>euc_sp<span class="ot">&lt;-</span><span class="fu">st_as_sf</span>(euc,<span class="at">coords=</span><span class="fu">c</span>(<span class="st">&quot;Long&quot;</span>, <span class="st">&quot;Lat&quot;</span>))</span>
<span id="cb383-2"><a href="geassoc.html#cb383-2" aria-hidden="true" tabindex="-1"></a><span class="fu">st_crs</span>(euc_sp) <span class="ot">&lt;-</span> <span class="fu">st_crs</span>(coast_cropped) </span>
<span id="cb383-3"><a href="geassoc.html#cb383-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb383-4"><a href="geassoc.html#cb383-4" aria-hidden="true" tabindex="-1"></a>mapg <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb383-5"><a href="geassoc.html#cb383-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> coast_cropped) <span class="sc">+</span></span>
<span id="cb383-6"><a href="geassoc.html#cb383-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> countries_cropped, <span class="at">fill =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb383-7"><a href="geassoc.html#cb383-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> states_cropped, <span class="at">fill =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb383-8"><a href="geassoc.html#cb383-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>euc,<span class="fu">aes</span>(<span class="at">x=</span>Long,<span class="at">y=</span>Lat,<span class="at">color=</span>Vulnerability),<span class="at">pch=</span><span class="dv">15</span>,<span class="at">size=</span>.<span class="dv">5</span>) <span class="sc">+</span></span>
<span id="cb383-9"><a href="geassoc.html#cb383-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_gradientn</span>(<span class="at">colours =</span> <span class="fu">Futcolors</span>(<span class="dv">100</span>)) <span class="sc">+</span></span>
<span id="cb383-10"><a href="geassoc.html#cb383-10" aria-hidden="true" tabindex="-1"></a>   <span class="fu">geom_sf</span>(<span class="at">data=</span>wifl_shape,<span class="at">fill=</span><span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb383-11"><a href="geassoc.html#cb383-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() </span>
<span id="cb383-12"><a href="geassoc.html#cb383-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb383-13"><a href="geassoc.html#cb383-13" aria-hidden="true" tabindex="-1"></a>mapg</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:unnamed-chunk-211"></span>
<img src="merida-workshop-2022_files/figure-html/unnamed-chunk-211-1.png" alt="Heat map of genomic vulnerability under the 2050 2.6 emission future prediction" width="672" />
<p class="caption">
FIGURE 7.8: Heat map of genomic vulnerability under the 2050 2.6 emission future prediction
</p>
</div>
<p><strong>How does this look if we plot the second scenario of higher gas emissions? Compare and contrast the two scenarios, keeping in mind that the colors are scaled within each scenario.</strong></p>
<div class="sourceCode" id="cb384"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb384-1"><a href="geassoc.html#cb384-1" aria-hidden="true" tabindex="-1"></a>futuregrid2 <span class="ot">&lt;-</span> <span class="fu">cbind</span>(future2[,<span class="fu">c</span>(<span class="st">&quot;X&quot;</span>,<span class="st">&quot;Y&quot;</span>)], <span class="fu">predict</span>(wiflforest,future2[,predictors_uncor]))</span>
<span id="cb384-2"><a href="geassoc.html#cb384-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb384-3"><a href="geassoc.html#cb384-3" aria-hidden="true" tabindex="-1"></a>coords<span class="ot">&lt;-</span>birdgrid[,<span class="fu">c</span>(<span class="st">&quot;X&quot;</span>,<span class="st">&quot;Y&quot;</span>)]</span>
<span id="cb384-4"><a href="geassoc.html#cb384-4" aria-hidden="true" tabindex="-1"></a>euc2 <span class="ot">&lt;-</span> <span class="fu">matrix</span>(<span class="at">data=</span><span class="cn">NA</span>,<span class="at">nrow=</span><span class="fu">nrow</span>(futuregrid2),<span class="at">ncol=</span><span class="dv">3</span>)</span>
<span id="cb384-5"><a href="geassoc.html#cb384-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(j <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">nrow</span>(currentgrid)) {</span>
<span id="cb384-6"><a href="geassoc.html#cb384-6" aria-hidden="true" tabindex="-1"></a>  euc2[j,] <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">as.numeric</span>(coords[j,]), </span>
<span id="cb384-7"><a href="geassoc.html#cb384-7" aria-hidden="true" tabindex="-1"></a>                <span class="fu">as.numeric</span>(<span class="fu">dist</span>(<span class="fu">rbind</span>(currentgrid[j,], futuregrid2[j,]))))</span>
<span id="cb384-8"><a href="geassoc.html#cb384-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb384-9"><a href="geassoc.html#cb384-9" aria-hidden="true" tabindex="-1"></a>euc2 <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(euc2)</span>
<span id="cb384-10"><a href="geassoc.html#cb384-10" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(euc2) <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">&quot;Long&quot;</span>,<span class="st">&quot;Lat&quot;</span>,<span class="st">&quot;Vulnerability&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb385"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb385-1"><a href="geassoc.html#cb385-1" aria-hidden="true" tabindex="-1"></a>euc_sp2<span class="ot">&lt;-</span><span class="fu">st_as_sf</span>(euc2,<span class="at">coords=</span><span class="fu">c</span>(<span class="st">&quot;Long&quot;</span>, <span class="st">&quot;Lat&quot;</span>))</span>
<span id="cb385-2"><a href="geassoc.html#cb385-2" aria-hidden="true" tabindex="-1"></a><span class="fu">st_crs</span>(euc_sp2) <span class="ot">&lt;-</span> <span class="fu">st_crs</span>(coast_cropped) </span>
<span id="cb385-3"><a href="geassoc.html#cb385-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb385-4"><a href="geassoc.html#cb385-4" aria-hidden="true" tabindex="-1"></a>mapg2 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb385-5"><a href="geassoc.html#cb385-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> coast_cropped) <span class="sc">+</span></span>
<span id="cb385-6"><a href="geassoc.html#cb385-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> countries_cropped, <span class="at">fill =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb385-7"><a href="geassoc.html#cb385-7" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data =</span> states_cropped, <span class="at">fill =</span> <span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb385-8"><a href="geassoc.html#cb385-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>euc2,<span class="fu">aes</span>(<span class="at">x=</span>Long,<span class="at">y=</span>Lat,<span class="at">color=</span>Vulnerability), <span class="at">pch=</span><span class="dv">15</span>,<span class="at">size=</span><span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb385-9"><a href="geassoc.html#cb385-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_colour_gradientn</span>(<span class="at">colours =</span> <span class="fu">Futcolors</span>(<span class="dv">100</span>))<span class="sc">+</span></span>
<span id="cb385-10"><a href="geassoc.html#cb385-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_sf</span>(<span class="at">data=</span>wifl_shape,<span class="at">fill=</span><span class="cn">NA</span>) <span class="sc">+</span></span>
<span id="cb385-11"><a href="geassoc.html#cb385-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_bw</span>() </span>
<span id="cb385-12"><a href="geassoc.html#cb385-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb385-13"><a href="geassoc.html#cb385-13" aria-hidden="true" tabindex="-1"></a>mapg2</span></code></pre></div>
<div class="figure"><span style="display:block;" id="fig:unnamed-chunk-213"></span>
<img src="merida-workshop-2022_files/figure-html/unnamed-chunk-213-1.png" alt="Heat map of genomic vulnerability under the 2050 8.5 emission future prediction" width="672" />
<p class="caption">
FIGURE 7.9: Heat map of genomic vulnerability under the 2050 8.5 emission future prediction
</p>
</div>
</div>
<div id="next-steps" class="section level2" number="7.7">
<h2><span class="header-section-number">7.7</span> Next steps</h2>
<div id="randomization" class="section level3" number="7.7.1">
<h3><span class="header-section-number">7.7.1</span> Randomization</h3>
<p>So as we’ve seen in this tutorial, grandient forest is a great way to understand which variables are important drivers of the population genomic structure of your species and with plotting the assoiciations in space, a great way to look at local adpation of different populations to different environmental variables. To understand how important, whether local adaptation is significant in your species, you need to run a randomization step. In this step, you are breaking up the actual associations of allele frequencies and enviornmnetal variables, randomizing the environmental predictors and running gradient forest again <em>multiple times</em>.</p>
<p>This code below is not evaluated because it takes a very long time to run one thousand randomizations.</p>
<pre><code>realtotal=wiflforest$species.pos.rsq
realaverage=sum(wiflforest$result)/realtotal
realaverage

for (i in 1:1000){

predsR=Ewifl[sample(nrow(Ewifl)),]

wiflforestR=gradientForest(cbind(predsR,Gwifl), predictor.vars=colnames(predsR), response.vars=colnames(Gwifl), ntree=10, transform = NULL, compact=F, maxLevel=1,trace=F,nbin=101)

 randtotal=wiflforestR$species.pos.rsq
 randaverage=sum(wiflforestR$result)/randtotal

 #totaldiff=realtotal-randtotal
 #avrdiff=realaverage-randaverage

 write.table(randtotal,file=paste(&quot;stored_results/6.0/WIFL.randtotal&quot;,sep=&quot;&quot;),row.names=FALSE,col.names=FALSE,append=T)
 write.table(randaverage,file=paste(&quot;stored_results/6.0/WIFL.randaverage&quot;,sep=&quot;&quot;),row.names=FALSE,col.names=FALSE,append=T)
}

randa &lt;- read_delim(&quot;stored_results/6.0/WIFL.randaverage&quot;,delim=&quot;\t&quot;,col_names = F)
randt=read.table(&quot;stored_results/6.0/WIFL.randtotal&quot;,delim=&quot;\t&quot;, col.names = F)
quantile(randa$V1,c(0.975))
#randa=as.numeric(randa)
#randt=as.numeric(randt)

library(ggpubr)
pdf(&quot;wiflforest.1000regtrees.hist.pdf&quot;)
par(mfrow=c(2,1))
hist(randa$V1,xlim=c(0.001,0.5), main=&quot;Average R-squared of Random Gradient Forests&quot;,xlab=&quot;Average R-squared of 10,000 WIFL SNPs&quot;)
abline(v=0.2006884,col=&quot;red&quot;) #replace with the 95th percential of the average R squared from real data
dev.off()
</code></pre>
<p>If your real results fall right in the middle of the randomized results, then even though you have ranked environmental variables underlying the population structure of your species, it’s no different than random, and therefore not significant.</p>
</div>
<div id="candidate-variants" class="section level3" number="7.7.2">
<h3><span class="header-section-number">7.7.2</span> Candidate variants</h3>
<p>If you find that gradient forest is significant, your next step might be wanting to identify variants associated with these top variables. Grandient forest is not really the way to do this. Rather, there are two programs that you can use: <em>Latent Factor Mixed Models (LFMM)</em> and <em>Redundancy Analyses (RDA)</em>.</p>
<ul>
<li>LFMM can identify candidate loci associated with top ranking environmental variables</li>
</ul>
<p><a href="https://bcm-uga.github.io/lfmm/" class="uri">https://bcm-uga.github.io/lfmm/</a></p>
<ul>
<li>RDA is a method to extract and summarise the variation in a set of response variables that can be explained by a set of explanatory variables. More accurately, RDA is a direct gradient analysis technique which summarises linear relationships between components of response variables that are “redundant” with (i.e. “explained” by) a set of explanatory variables. Very similar to gradient forest, but also can identify candidate genes associated with environmental variables.</li>
</ul>
<p><a href="https://popgen.nescent.org/2018-03-27_RDA_GEA.html" class="uri">https://popgen.nescent.org/2018-03-27_RDA_GEA.html</a></p>

</div>
</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="spatdata.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="conclude.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook/js/app.min.js"></script>
<script src="libs/gitbook/js/clipboard.min.js"></script>
<script src="libs/gitbook/js/plugin-search.js"></script>
<script src="libs/gitbook/js/plugin-sharing.js"></script>
<script src="libs/gitbook/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook/js/plugin-bookdown.js"></script>
<script src="libs/gitbook/js/jquery.highlight.js"></script>
<script src="libs/gitbook/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": true,
"facebook": false,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"whatsapp": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": "https://github.com/eriqande/merida-workshop-2022/edit/master/6.0-genetic-environment-associations.Rmd",
"text": "Edit"
},
"history": {
"link": "https://github.com/eriqande/merida-workshop-2022/commits/master/6.0-genetic-environment-associations.Rmd",
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["merida-workshop-2022.pdf", "merida-workshop-2022.epub"],
"search": {
"engine": "fuse",
"options": null
},
"toc": {
"collapse": "section"
}
});
});
</script>

</body>

</html>
